{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ee0e534d",
   "metadata": {},
   "source": [
    "# Sample Mass-Difference Networks in Metabolomics Data Analysis\n",
    "\n",
    "Notebook to support the study on the application of **Sample M**ass-**Di**fference **N**etworks as a highly specific competing form of pre-processing procedure for high-resolution metabolomics data.\n",
    "\n",
    "Mass-Difference Networks are focused into making networks from a list of masses. Each _m/z_ will represent a node. Nodes will be connected if the difference in their masses can be associated to a simple chemical reaction (enzymatic or non-enzymatic) that led to a change in the elemental composition of its metabolite.\n",
    "\n",
    "The set of mass differences used to build said networks are called a set of MDBs - Mass-Difference-based Building block.\n",
    "\n",
    "This is notebook `paper_sMDiNs_database_prep.ipynb`.\n",
    "\n",
    "\n",
    "## Organization of the Notebook\n",
    "\n",
    "- **Set up database of the 7 benchmark datasets**\n",
    "- Description of the benchmark datasets\n",
    "- **Building transformation list (of MDBs) and setting up list of masses files to use in Cytoscape's MetaNetter (to build MDiNs)**\n",
    "- Show the benchmark datasets characteristics\n",
    "- Explanation of the parameters used in Cytoscape's MetaNetter 2.0 to build the MDiNs.\n",
    "- Apply intensity-based pre-treatments.\n",
    "- **Store pre-processed and pre-treated data**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c85e0c49",
   "metadata": {},
   "source": [
    "#### Needed Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "44640c11",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas.testing import assert_frame_equal\n",
    "\n",
    "# Metabolinks package\n",
    "import metabolinks as mtl\n",
    "import metabolinks.transformations as transf\n",
    "\n",
    "from pathlib import Path\n",
    "from sklearn.experimental import enable_iterative_imputer\n",
    "from sklearn.impute import IterativeImputer\n",
    "from sklearn.tree import DecisionTreeRegressor\n",
    "from sklearn.ensemble import ExtraTreesRegressor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ffe409ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9355b53f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# json for persistence\n",
    "import json\n",
    "from time import perf_counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e01cea88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Atomic masses - https://ciaaw.org/atomic-masses.htm\n",
    "#Isotopic abundances-https://ciaaw.org/isotopic-abundances.htm/https://www.degruyter.com/view/journals/pac/88/3/article-p293.xml\n",
    "# Isotopic abundances from Pure Appl. Chem. 2016; 88(3): 293–306,\n",
    "# Isotopic compositions of the elements 2013 (IUPAC Technical Report), doi: 10.1515/pac-2015-0503\n",
    "\n",
    "chemdict = {'H':(1.0078250322, 0.999844),\n",
    "            'C':(12.000000000, 0.988922),\n",
    "            'N':(14.003074004, 0.996337),\n",
    "            'O':(15.994914619, 0.9976206),\n",
    "            'Na':(22.98976928, 1.0),\n",
    "            'P':(30.973761998, 1.0),\n",
    "            'S':(31.972071174, 0.9504074),\n",
    "            'Cl':(34.9688527, 0.757647),\n",
    "            'F':(18.998403163, 1.0),\n",
    "            'C13':(13.003354835, 0.011078) # Carbon 13 isotope\n",
    "           } \n",
    "\n",
    "# electron mass from NIST http://physics.nist.gov/cgi-bin/cuu/Value?meu|search_for=electron+mass\n",
    "electron_mass = 0.000548579909065"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b59f4d72",
   "metadata": {},
   "source": [
    "## Description of dataset records\n",
    "\n",
    "`datasets` is the global dict that holds all data sets. It is a **dict of dict's**.\n",
    "\n",
    "Each data set is **represented as a dict**.\n",
    "\n",
    "Each record has the following fields (keys):\n",
    "\n",
    "- `name`: the table/figure name of the data set\n",
    "- `source`: the biological source for each dataset\n",
    "- `mode`: the aquisition mode\n",
    "- `alignment`: the alignment used to generate the data matrix\n",
    "- `data`: the data matrix\n",
    "- `target`: the sample labels, possibly already integer encoded\n",
    "- `MDiN`: Mass-Difference Network - Not present here, only on sMDiNsAnalysis notebook\n",
    "- `<treatment name>`: transformed data matrix / network. These treatment names can be\n",
    "    - `original`: an alias to `data`\n",
    "    - `Ionly`: missing value imputed data by 1/5 of the minimum value in each sample in the dataset, only\n",
    "    - `P`: Pareto scaled data\n",
    "    - `NP`: Pareto scaled and normalized\n",
    "    - `NGP`: normalized, glog transformed and Pareto scaled\n",
    "    - `Ionly_RF`: missing value imputed data by random forests, only\n",
    "    - `P_RF`: Pareto scaled data\n",
    "    - `NP_RF`: Pareto scaled and normalized\n",
    "    - `NGP_RF`: normalized, glog transformed and Pareto scaled\n",
    "    - `IDT`: `NGP_RF` or `NGP` - Intensity-based Data pre-Treatment chosen as comparison based on which of the two performed better for each dataset and each statistical method\n",
    "    - `sMDiN`: Sample Mass-Difference Networks - Not present here, only on sMDiNsAnalysis notebook\n",
    "       \n",
    "- `<sMDiN analysis name>`: data matrix from nework analysis of MDiNs - Not in this notebook\n",
    "    - `Degree`: degree analysis of each sMDiN\n",
    "    - `Betweenness`: betweenness centrality analysis of each sMDiN\n",
    "    - `Closeness`: closeness centrality of analysis of each sMDiN\n",
    "    - `MDBI`: analysis on the impact of each MDB (Mass-Difference based building-block) on building each sMDiN\n",
    "    - `GCD11`: Graphlet Correlation Distance of 11 different orbits (maximum of 4-node graphlets) between each sMDiN.\n",
    "\n",
    "The keys of `datasets` may be shared with dicts holding records resulting from comparison analysis.\n",
    "\n",
    "Here are the keys (and respective names) of datasets used in this study:\n",
    "\n",
    "- GD_neg_global2 (GDg2-)\n",
    "- GD_pos_global2 (GDg2+)\n",
    "- GD_neg_class2 (GDc2-)\n",
    "- GD_pos_class2 (GDc2+)\n",
    "- YD (YD)\n",
    "- vitis_types (GD types)\n",
    "- HD (HD)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "503a60ea",
   "metadata": {},
   "source": [
    "### Description of grapevine data sets\n",
    "\n",
    "Grapevine Datasets (Negative and Positive) - 33 samples belonging to 11 different grapevine varieties (3 samples per variety/biological group) of FT-ICR-MS metabolomics data obtained in negative and positive ionization mode.\n",
    "\n",
    "5 different _Vitis_ species (other than _V. vinifera_) varieties:\n",
    "\n",
    "- CAN - 3 Samples (14, 15, 16) of _V. candicans Engelmann_ (VIVC variety number: 13508)\n",
    "- RIP - 3 Samples (17, 18, 19) of _V. riparia Michaux_ (Riparia Gloire de Montpellier, VIVC variety number: 4824) \n",
    "- ROT - 3 Samples (20, 21, 22) of _V. rotundifolia_ (Muscadinia Rotundifolia Michaux cv. Rotundifolia, VIVC variety number: 13586)\n",
    "- RU - 3 Samples (35, 36, 37) of _V. rupestris Scheele_ (Rupestris du lot, VIVC variety number: 10389)\n",
    "- LAB - 3 Samples (8, 9, 10) of _V. labrusca_ (Isabella, VIVC variety number: 5560)\n",
    "\n",
    "6 different _V. vinifera_ cultivars varieties are:\n",
    "\n",
    "- SYL - 3 samples (11, 12, 13) of the subspecies _sylvestris_ (VIVC variety number: -)\n",
    "- CS - 3 Samples (29, 30, 31) of the subspecies _sativa_ cultivar Cabernet Sauvignon (VIVC variety number: 1929)\n",
    "- PN - 3 Samples (23, 24, 25) of the subspecies _sativa_ cultivar Pinot Noir (VIVC variety number: 9279)\n",
    "- REG - 3 Samples (38, 39, 40) of the subspecies _sativa_ cultivar Regent (VIVC variety number: 4572)\n",
    "- RL - 3 Samples (26, 27, 28) of the subspecies _sativa_ cultivar Riesling Weiss (VIVC variety number: 10077)\n",
    "- TRI - 3 Samples (32, 33, 34) of the subspecies _sativa_ cultivar Cabernet Sauvignon (VIVC variety number: 15685)\n",
    "\n",
    "Data acquired by Maia et al. (2020):\n",
    "\n",
    "- Maia M, Ferreira AEN, Nascimento R, et al. Integrating metabolomics and targeted gene expression to uncover potential biomarkers of fungal / oomycetes ‑ associated disease susceptibility in grapevine. Sci Rep. Published online 2020:1-15. doi:10.1038/s41598-020-72781-2\n",
    "- Maia M, Figueiredo A, Silva MS, Ferreira A. Grapevine untargeted metabolomics to uncover potential biomarkers of fungal/oomycetes-associated diseases. 2020. doi:10.6084/m9.figshare.12357314.v1\n",
    "\n",
    "**Peak Alignment** and **Peak Filtering** were performed with function `metabolinks.peak_alignment.align()`. Human leucine enkephalin (Sigma Aldrich) was used as the reference feature (internal standard, [M+H]+ = 556.276575 Da or [M-H]- = 554.262022 Da).\n",
    "\n",
    "**4** data matrices were constructed from this data:\n",
    "\n",
    "- Data sets named `GD_pos_global2` (GDg2+) and `GD_neg_global2` (GDg2-) were generated after retaining only features that occur (globally) at least twice in all 33 samples of the data sets (filtering/alignment) for the **positive mode** data acquisition and the **negative mode** data acquisition, respectively.\n",
    "- Data sets named `GD_pos_class2` (GDc2+) and `GD_neg_class2` (GDc2-) were generated after retaining only features that occur at least twice in the three replicates of at least one _Vitis_ variety in the data sets (filtering/alignment) for the **positive mode** data acquisition and the **negative mode** data acquisition, respectively.\n",
    "\n",
    "For the purpose of assessing the performance of supervised methods each of these four datasets was used with target labels defining classes corresponding to replicates of each of the 11 Vitis species/cultivars.\n",
    "\n",
    "For the purpose of assessing the performance of supervised methods under a binary (two-class) problem, data set `GD_neg_class2` was also used with target labels defining two classes: Vitis vinifera cultivars and \"wild\", non-vinifera Vitis species. This is dataset `vitis_types` (GD types)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc74ce5b",
   "metadata": {},
   "source": [
    "### Description of the yeast data set\n",
    "\n",
    "Yeast dataset - 15 samples belonging to 5 different yeast strains of _Saccharomyces cerevisiae_ (3 biological replicates per strain/biological group) of FT-ICR-MS metabolomics data obtained in positive ionization mode. The 5 strains were: the reference strain BY4741 (represented as BY) and 4 single-gene deletion mutants of this strain – ΔGLO1 (GLO1), ΔGLO2 (GLO2), ΔGRE3 (GRE3) and ΔENO1 (ENO1). These deleted genes are directly or indirectly related to methylglyoxal metabolism.\n",
    "\n",
    "Data acquired by Luz et al. (2021):\n",
    "\n",
    "- Luz J, Pendão AS, Silva MS, Cordeiro C. FT-ICR-MS based untargeted metabolomics for the discrimination of yeast mutants. 2021. doi:10.6084/m9.figshare.15173559.v1\n",
    "\n",
    "**Peak Alignment** and **Peak Filtering** was performed with MetaboScape 4.0 software (see reference above for details in sample preparation, pre-processing, formula assignment). In short, Yeast Dataset was obtained with Electrospray Ionization in Positive Mode and pre-processed by MetaboScape 4.0 (Bruker Daltonics). Human leucine enkephalin (Sigma Aldrich) was used as the reference feature (internal standard, [M+H]+ = 556.276575 Da or [M-H]- = 554.262022 Da).\n",
    "\n",
    "A data matrix was constructed from this data: data set named **`YD` (YD)** was generated after retaining only features that occur (globally) at least twice in all 15 samples (filtering/alignment).\n",
    "\n",
    "Furthermore, peaks higher than 1000 m/z were filtered out. Formula assignment was performed with MetaboScape 4.0 software and, to consider peaks that had the same formula assigned, these were joined together in a single peak (`feature_filter` function).\n",
    "\n",
    "For the purpose of assessing the performance of supervised methods, this data set was used with target labels defining classes corresponding to replicates of each of the 4 yeast strains."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c1c1f11",
   "metadata": {},
   "source": [
    "#### Loading yeast data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d193d85c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_MetScape_file(filename,\n",
    "                          col_renamer=None,\n",
    "                          add_labels=None,\n",
    "                          remove_ref_feat=None,):\n",
    "    \n",
    "    \"\"\"Read in a MetaboScape bucket table from a CSV file.\"\"\"\n",
    "    \n",
    "    data = pd.read_csv(filename).set_index('Bucket label')\n",
    "    \n",
    "    # optionally rename sample_names\n",
    "    if col_renamer is not None:\n",
    "        data = data.rename(columns=renamer)\n",
    "    \n",
    "    # optionally remove a reference feature (if already normalized)\n",
    "    if remove_ref_feat is not None:\n",
    "        #print(f'Feature {remove_ref_feat}\\n{data.loc[remove_ref_feat, :]}\\n----------')\n",
    "        data = data.drop(index=[remove_ref_feat])\n",
    "        \n",
    "    # Joining rows that have the same formula assigned (with feature_filter) while keeping rows without formulas assigned.\n",
    "    data = feature_filter(data)\n",
    "    \n",
    "    # split in peak metadata and intensities\n",
    "    peak_cols = ['m/z', 'Name', 'Formula']\n",
    "    intensity_cols = [c for c in list(data.columns) if c not in peak_cols]\n",
    "    peaks = data[peak_cols]\n",
    "    intensities = data[intensity_cols]\n",
    "\n",
    "    # replace zeros for NaN's\n",
    "    intensities = intensities.replace(0, np.nan).dropna(how='all')\n",
    "    \n",
    "    # force peaks to have the same features as the (trimmed) intensities\n",
    "    peaks = peaks.reindex(intensities.index)\n",
    "\n",
    "    # optionally, add labels to intensities\n",
    "    if add_labels is not None:\n",
    "        intensities = mtl.add_labels(intensities, labels=add_labels)\n",
    "    \n",
    "    return {'filename': filename,\n",
    "            'peaks':peaks,\n",
    "            'intensities': intensities}\n",
    "\n",
    "def renamer(colname):\n",
    "    # Util to optionally remove all those 00000 from sample names\n",
    "    return ''.join(colname.split('00000'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e5967446",
   "metadata": {},
   "outputs": [],
   "source": [
    "def feature_filter(Spectra):\n",
    "    \"\"\"Join all features that share the same formula by addition.\n",
    "       \n",
    "       Spectra: Pandas DataFrame (dataset).\n",
    "       \n",
    "       Returns: Pandas DataFrame with filtered dataset.\n",
    "    \"\"\"\n",
    "\n",
    "    df = Spectra\n",
    "    # Variables to store results\n",
    "    rp = pd.DataFrame(columns = df.columns)\n",
    "    rep = []\n",
    "    # Series with number of times each formula appears in the dataset\n",
    "    form_count = df['Formula'].value_counts()\n",
    "    a = 0\n",
    "    \n",
    "    for i in df['Formula']:\n",
    "        if pd.isnull(i) == True: # If no formula is assigned to the feature, the feature stays in the new dataset\n",
    "            rp = rp.append(df.iloc[a,:])\n",
    "\n",
    "        else: # If a formula is assigned\n",
    "            if form_count[i] > 1: # If the formula appears more than one time in the dataset\n",
    "                if i not in rep: # And If it didn't already appear before\n",
    "                    \n",
    "                    # Bucket label, name and formula info are the ones from the greater intensity m/z peak (of peaks with the\n",
    "                    # same formula)\n",
    "                    peaks = df.loc[df['Formula'] == i,:].iloc[:,3:].sum(axis = 1)\n",
    "                    peaks_idx = peaks.idxmax()\n",
    "                    info = df.loc[peaks_idx].iloc[:3] # Bucket label, name and formula info\n",
    "                    \n",
    "                    # Addition of all m/z peaks intensity with the same formula\n",
    "                    newrow = df.loc[df['Formula'] == i,:].iloc[:,3:].sum(axis = 0)\n",
    "                    newrow_c = pd.concat([info, newrow], axis = 0) # Join with the m/z peak metadata\n",
    "                    # m/z peak is from the greater intensity m/z peak (of peaks with the same formula)\n",
    "                    newrow_c = newrow_c.rename(df.loc[peaks_idx].name)\n",
    "                    \n",
    "                    # Append peak with intensity added from all peaks sharing the same formula\n",
    "                    rp = rp.append(newrow_c)\n",
    "\n",
    "                    rep.append(i) # Append repeating formulas so this process isn't repeated again for the other m/z peaks\n",
    "\n",
    "            else:\n",
    "                rp = rp.append(df.iloc[a,:]) # Append peaks with formulas that aren't repeated in the dataset\n",
    "        a = a + 1\n",
    "\n",
    "    # return DataFrame\n",
    "    return rp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cb2457fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labels of the 5 biological groups (5 yeast strains) - only added after\n",
    "yeast_classes = 'WT ΔGRE3 ΔENO1 ΔGLO1 ΔGLO2'.split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6270c542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the file and keep results in dicts \n",
    "\n",
    "prefix_to_drop = None # change to 'ENO' to remove ENO strain\n",
    "\n",
    "# MetaboScape data\n",
    "MS_data = read_MetScape_file('5yeasts_notnorm.csv', \n",
    "                             remove_ref_feat=None,\n",
    "                             add_labels=None,\n",
    "                             col_renamer=renamer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a065973a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>307.083818</th>\n",
       "      <th>555.269298</th>\n",
       "      <th>624.087308</th>\n",
       "      <th>493.316816</th>\n",
       "      <th>257.102875</th>\n",
       "      <th>780.109207</th>\n",
       "      <th>347.063081</th>\n",
       "      <th>254.224610</th>\n",
       "      <th>625.086929</th>\n",
       "      <th>663.109142</th>\n",
       "      <th>...</th>\n",
       "      <th>734.677027</th>\n",
       "      <th>648.574264</th>\n",
       "      <th>634.548871</th>\n",
       "      <th>675.482247</th>\n",
       "      <th>739.516465</th>\n",
       "      <th>916.507669</th>\n",
       "      <th>550.163642</th>\n",
       "      <th>914.598352</th>\n",
       "      <th>802.744007</th>\n",
       "      <th>535.097415</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BY0_1</th>\n",
       "      <td>391042880.0</td>\n",
       "      <td>111270160.0</td>\n",
       "      <td>130088936.0</td>\n",
       "      <td>1.122515e+08</td>\n",
       "      <td>54234476.0</td>\n",
       "      <td>33863468.0</td>\n",
       "      <td>35209136.0</td>\n",
       "      <td>9935943.0</td>\n",
       "      <td>25313166.0</td>\n",
       "      <td>21631346.0</td>\n",
       "      <td>...</td>\n",
       "      <td>726365.5625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BY0_2</th>\n",
       "      <td>398003680.0</td>\n",
       "      <td>110423984.0</td>\n",
       "      <td>127133608.0</td>\n",
       "      <td>1.139838e+08</td>\n",
       "      <td>53720908.0</td>\n",
       "      <td>32622178.0</td>\n",
       "      <td>35636648.0</td>\n",
       "      <td>17679414.0</td>\n",
       "      <td>24989162.0</td>\n",
       "      <td>24420090.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BY0_3</th>\n",
       "      <td>399481952.0</td>\n",
       "      <td>111225424.0</td>\n",
       "      <td>131208568.0</td>\n",
       "      <td>1.158259e+08</td>\n",
       "      <td>55826148.0</td>\n",
       "      <td>34655496.0</td>\n",
       "      <td>37943456.0</td>\n",
       "      <td>19530026.0</td>\n",
       "      <td>24684334.0</td>\n",
       "      <td>24862190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>666204.7500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>558134.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_1</th>\n",
       "      <td>270906400.0</td>\n",
       "      <td>92849040.0</td>\n",
       "      <td>130682664.0</td>\n",
       "      <td>1.104378e+08</td>\n",
       "      <td>46503348.0</td>\n",
       "      <td>33206608.0</td>\n",
       "      <td>32640452.0</td>\n",
       "      <td>5713388.5</td>\n",
       "      <td>25890872.0</td>\n",
       "      <td>28692028.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>613780.4375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_2</th>\n",
       "      <td>271023520.0</td>\n",
       "      <td>94888320.0</td>\n",
       "      <td>130844488.0</td>\n",
       "      <td>1.144803e+08</td>\n",
       "      <td>44828256.0</td>\n",
       "      <td>34597444.0</td>\n",
       "      <td>31696134.0</td>\n",
       "      <td>10139615.0</td>\n",
       "      <td>25548978.0</td>\n",
       "      <td>26964150.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>579154.75</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_3</th>\n",
       "      <td>274854272.0</td>\n",
       "      <td>94514272.0</td>\n",
       "      <td>130896568.0</td>\n",
       "      <td>1.233897e+08</td>\n",
       "      <td>46361588.0</td>\n",
       "      <td>33131336.0</td>\n",
       "      <td>31668180.0</td>\n",
       "      <td>16434200.0</td>\n",
       "      <td>25518902.0</td>\n",
       "      <td>26654308.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_1</th>\n",
       "      <td>255276736.0</td>\n",
       "      <td>150620544.0</td>\n",
       "      <td>136612464.0</td>\n",
       "      <td>3.220034e+07</td>\n",
       "      <td>72465536.0</td>\n",
       "      <td>35298204.0</td>\n",
       "      <td>31236114.0</td>\n",
       "      <td>10759292.0</td>\n",
       "      <td>25798230.0</td>\n",
       "      <td>22221992.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>720792.625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>538023.5625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_2</th>\n",
       "      <td>251610656.0</td>\n",
       "      <td>148516304.0</td>\n",
       "      <td>134972800.0</td>\n",
       "      <td>3.422184e+07</td>\n",
       "      <td>70376640.0</td>\n",
       "      <td>35198252.0</td>\n",
       "      <td>30789652.0</td>\n",
       "      <td>25896608.0</td>\n",
       "      <td>26849502.0</td>\n",
       "      <td>23433602.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>626384.625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>609298.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_3</th>\n",
       "      <td>251024624.0</td>\n",
       "      <td>150645840.0</td>\n",
       "      <td>135922432.0</td>\n",
       "      <td>3.370513e+07</td>\n",
       "      <td>68912816.0</td>\n",
       "      <td>34380756.0</td>\n",
       "      <td>30916958.0</td>\n",
       "      <td>29723808.0</td>\n",
       "      <td>26606410.0</td>\n",
       "      <td>23500090.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>655682.1875</td>\n",
       "      <td>624387.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>561889.8750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_1</th>\n",
       "      <td>283822048.0</td>\n",
       "      <td>202869216.0</td>\n",
       "      <td>130259064.0</td>\n",
       "      <td>4.167029e+07</td>\n",
       "      <td>58269456.0</td>\n",
       "      <td>32095004.0</td>\n",
       "      <td>29669478.0</td>\n",
       "      <td>26538900.0</td>\n",
       "      <td>26013280.0</td>\n",
       "      <td>21245436.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>548121.4375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_2</th>\n",
       "      <td>285664416.0</td>\n",
       "      <td>205537504.0</td>\n",
       "      <td>128839264.0</td>\n",
       "      <td>4.418028e+07</td>\n",
       "      <td>56775068.0</td>\n",
       "      <td>32852638.0</td>\n",
       "      <td>30652686.0</td>\n",
       "      <td>27920220.0</td>\n",
       "      <td>24769882.0</td>\n",
       "      <td>21539278.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>579864.500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>546345.2500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_3</th>\n",
       "      <td>295673952.0</td>\n",
       "      <td>209370176.0</td>\n",
       "      <td>132857032.0</td>\n",
       "      <td>4.338860e+07</td>\n",
       "      <td>58019620.0</td>\n",
       "      <td>35047176.0</td>\n",
       "      <td>30938974.0</td>\n",
       "      <td>29449846.0</td>\n",
       "      <td>26228898.0</td>\n",
       "      <td>21589446.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_1</th>\n",
       "      <td>287722656.0</td>\n",
       "      <td>421969440.0</td>\n",
       "      <td>131498984.0</td>\n",
       "      <td>6.056658e+07</td>\n",
       "      <td>59549848.0</td>\n",
       "      <td>33618800.0</td>\n",
       "      <td>34360160.0</td>\n",
       "      <td>9995730.0</td>\n",
       "      <td>25419526.0</td>\n",
       "      <td>26483844.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>683419.6250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_2</th>\n",
       "      <td>287749600.0</td>\n",
       "      <td>416559200.0</td>\n",
       "      <td>131380096.0</td>\n",
       "      <td>6.252698e+07</td>\n",
       "      <td>60346920.0</td>\n",
       "      <td>32398028.0</td>\n",
       "      <td>35621036.0</td>\n",
       "      <td>13618347.0</td>\n",
       "      <td>25224294.0</td>\n",
       "      <td>26875734.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>722504.8750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>713922.4375</td>\n",
       "      <td>663099.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>561280.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_3</th>\n",
       "      <td>290416512.0</td>\n",
       "      <td>419548864.0</td>\n",
       "      <td>131334040.0</td>\n",
       "      <td>6.181474e+07</td>\n",
       "      <td>61914296.0</td>\n",
       "      <td>33372098.0</td>\n",
       "      <td>36017192.0</td>\n",
       "      <td>19396972.0</td>\n",
       "      <td>25047738.0</td>\n",
       "      <td>25212104.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>597485.75</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 1893 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          307.083818   555.269298   624.087308    493.316816  257.102875  \\\n",
       "BY0_1    391042880.0  111270160.0  130088936.0  1.122515e+08  54234476.0   \n",
       "BY0_2    398003680.0  110423984.0  127133608.0  1.139838e+08  53720908.0   \n",
       "BY0_3    399481952.0  111225424.0  131208568.0  1.158259e+08  55826148.0   \n",
       "GRE3_1   270906400.0   92849040.0  130682664.0  1.104378e+08  46503348.0   \n",
       "GRE3_2   271023520.0   94888320.0  130844488.0  1.144803e+08  44828256.0   \n",
       "GRE3_3   274854272.0   94514272.0  130896568.0  1.233897e+08  46361588.0   \n",
       "ENO1_1   255276736.0  150620544.0  136612464.0  3.220034e+07  72465536.0   \n",
       "ENO1_2   251610656.0  148516304.0  134972800.0  3.422184e+07  70376640.0   \n",
       "ENO1_3   251024624.0  150645840.0  135922432.0  3.370513e+07  68912816.0   \n",
       "dGLO1_1  283822048.0  202869216.0  130259064.0  4.167029e+07  58269456.0   \n",
       "dGLO1_2  285664416.0  205537504.0  128839264.0  4.418028e+07  56775068.0   \n",
       "dGLO1_3  295673952.0  209370176.0  132857032.0  4.338860e+07  58019620.0   \n",
       "GLO2_1   287722656.0  421969440.0  131498984.0  6.056658e+07  59549848.0   \n",
       "GLO2_2   287749600.0  416559200.0  131380096.0  6.252698e+07  60346920.0   \n",
       "GLO2_3   290416512.0  419548864.0  131334040.0  6.181474e+07  61914296.0   \n",
       "\n",
       "         780.109207  347.063081  254.224610  625.086929  663.109142  ...  \\\n",
       "BY0_1    33863468.0  35209136.0   9935943.0  25313166.0  21631346.0  ...   \n",
       "BY0_2    32622178.0  35636648.0  17679414.0  24989162.0  24420090.0  ...   \n",
       "BY0_3    34655496.0  37943456.0  19530026.0  24684334.0  24862190.0  ...   \n",
       "GRE3_1   33206608.0  32640452.0   5713388.5  25890872.0  28692028.0  ...   \n",
       "GRE3_2   34597444.0  31696134.0  10139615.0  25548978.0  26964150.0  ...   \n",
       "GRE3_3   33131336.0  31668180.0  16434200.0  25518902.0  26654308.0  ...   \n",
       "ENO1_1   35298204.0  31236114.0  10759292.0  25798230.0  22221992.0  ...   \n",
       "ENO1_2   35198252.0  30789652.0  25896608.0  26849502.0  23433602.0  ...   \n",
       "ENO1_3   34380756.0  30916958.0  29723808.0  26606410.0  23500090.0  ...   \n",
       "dGLO1_1  32095004.0  29669478.0  26538900.0  26013280.0  21245436.0  ...   \n",
       "dGLO1_2  32852638.0  30652686.0  27920220.0  24769882.0  21539278.0  ...   \n",
       "dGLO1_3  35047176.0  30938974.0  29449846.0  26228898.0  21589446.0  ...   \n",
       "GLO2_1   33618800.0  34360160.0   9995730.0  25419526.0  26483844.0  ...   \n",
       "GLO2_2   32398028.0  35621036.0  13618347.0  25224294.0  26875734.0  ...   \n",
       "GLO2_3   33372098.0  36017192.0  19396972.0  25047738.0  25212104.0  ...   \n",
       "\n",
       "          734.677027   648.574264  634.548871   675.482247   739.516465  \\\n",
       "BY0_1    726365.5625          NaN         NaN          NaN          NaN   \n",
       "BY0_2            NaN          NaN         NaN          NaN          NaN   \n",
       "BY0_3    666204.7500          NaN         NaN          NaN          NaN   \n",
       "GRE3_1           NaN          NaN         NaN          NaN          NaN   \n",
       "GRE3_2           NaN          NaN         NaN          NaN          NaN   \n",
       "GRE3_3           NaN          NaN         NaN          NaN          NaN   \n",
       "ENO1_1           NaN          NaN  720792.625          NaN          NaN   \n",
       "ENO1_2           NaN          NaN         NaN          NaN          NaN   \n",
       "ENO1_3           NaN  655682.1875  624387.000          NaN  561889.8750   \n",
       "dGLO1_1          NaN          NaN         NaN          NaN          NaN   \n",
       "dGLO1_2          NaN          NaN         NaN          NaN          NaN   \n",
       "dGLO1_3          NaN          NaN         NaN          NaN          NaN   \n",
       "GLO2_1           NaN          NaN         NaN  683419.6250          NaN   \n",
       "GLO2_2           NaN  722504.8750         NaN  713922.4375  663099.1875   \n",
       "GLO2_3           NaN          NaN         NaN          NaN          NaN   \n",
       "\n",
       "         916.507669   550.163642   914.598352  802.744007   535.097415  \n",
       "BY0_1           NaN          NaN          NaN         NaN          NaN  \n",
       "BY0_2           NaN          NaN          NaN         NaN          NaN  \n",
       "BY0_3           NaN  558134.1875          NaN         NaN          NaN  \n",
       "GRE3_1          NaN  613780.4375          NaN         NaN          NaN  \n",
       "GRE3_2          NaN          NaN          NaN   579154.75          NaN  \n",
       "GRE3_3          NaN          NaN          NaN         NaN          NaN  \n",
       "ENO1_1          NaN          NaN          NaN         NaN  538023.5625  \n",
       "ENO1_2   626384.625          NaN  609298.9375         NaN          NaN  \n",
       "ENO1_3          NaN          NaN          NaN         NaN          NaN  \n",
       "dGLO1_1         NaN          NaN  548121.4375         NaN          NaN  \n",
       "dGLO1_2  579864.500          NaN          NaN         NaN  546345.2500  \n",
       "dGLO1_3         NaN          NaN          NaN         NaN          NaN  \n",
       "GLO2_1          NaN          NaN          NaN         NaN          NaN  \n",
       "GLO2_2          NaN          NaN          NaN   561280.00          NaN  \n",
       "GLO2_3          NaN          NaN          NaN   597485.75          NaN  \n",
       "\n",
       "[15 rows x 1893 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# keep features that appear in at least two samples\n",
    "yeast_datamatrix = transf.keep_atleast(MS_data['intensities'].transpose(), minimum=2)\n",
    "\n",
    "# Changing the index to be numbers - our list of neutral masses from the bucket lists.\n",
    "new_columns = []\n",
    "for i in range(len(yeast_datamatrix.columns)):\n",
    "    new_columns.append(float(yeast_datamatrix.columns[i][:-3]))\n",
    "yeast_datamatrix.columns = new_columns\n",
    "\n",
    "# Only keep peaks below 1000 m/z - Very few peaks above 1000 m/z.\n",
    "yeast_datamatrix = yeast_datamatrix.T[yeast_datamatrix.columns < 1000].T\n",
    "\n",
    "yeast_datamatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1678aa51",
   "metadata": {},
   "source": [
    "### Description of the Human dataset\n",
    "\n",
    "Human Dataset - 249 samples belonging to 2 different classes of HILIC-MS data obtained in positive ionization mode. The Chromatography system was a Thermo Dionex Ultimate 3000 with the column being a Waters Xbridge BEH HILIC (75 x 2.1mm, 2.5um). Mass Spectrometry was performed in a Thermo Q Exactive HF hybrid Orbitrap operating in positive electrospray ionization mode. Further information in the data deposition site mentioned below.\n",
    "\n",
    "The 2 classes are:\n",
    "\n",
    "- 135 pre-operative blood (serum) samples from patients with **'No Recurrence'** of Prostate Cancer after Radical Prostatectomy\n",
    "- 114 pre-operative blood (serum) samples from patients with **'Recurrence'** of Prostate Cancer after Radical Prostatectomy\n",
    "\n",
    "I think this means samples were collected before the Prostatectomy before knowledge of recurrence or no recurrence? There was a total of 80 subjects.\n",
    "\n",
    "Data taken from data deposition site: https://www.metabolomicsworkbench.org/data/DRCCMetadata.php?Mode=Study&DataMode=AllData&StudyID=ST001082&StudyType=MS&ResultType=5#DataTabs\n",
    "\n",
    "This data is available at the NIH Common Fund's National Metabolomics Data Repository (NMDR) website, the Metabolomics Workbench, https://www.metabolomicsworkbench.org where it has been assigned Project ID PR000724. The data can be accessed directly via it's Project DOI: 10.21228/M83D5V. \n",
    "\n",
    "Data deposited by the Georgia Institute of Technology and researcher Facundo Fernandez. This data was taken already aligned in a 2D numerical data matrix. \n",
    "\n",
    "Data used in the paper: Clendinen CS, Gaul DA, Monge ME, et al. Preoperative Metabolic Signatures of Prostate Cancer Recurrence Following Radical  Prostatectomy. J Proteome Res. 2019;18(3):1316-1327. doi:10.1021/acs.jproteome.8b00926\n",
    "\n",
    "Dataset named **HD (HD)** was generated after retaining only features that occur (globally) at least twice in all samples of the dataset. For the purpose of assessing the performance of supervised methods this dataset was used with target labels defining classes based on which of the 2 classes mentioned prior, the samples belonged to.\n",
    "\n",
    "\n",
    "The data read from the site seemed to be completely untreated, which was perfect for our efforts. We can't be sure but the presence of plenty of missing values (very useful for us) points to no great peak filtering and missing value imputation being made, the bar plot built from adding up all intensities of each sample seems to point that no normalization had yet been employed in the data, and the distribution observed in the boxplot of the intensities of some of the features in the samples seems to point that no scaling had been used yet in the data as well.\n",
    "\n",
    "The data also had 2 copies of each sample, where the 2nd copies were equal to the first ones multiplied by a constant. This constant was different for each sample. We removed the presence of the 2nd copies. The index (metabolites) were reduced from the notation m/z value_retention time to only have the m/z value.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a2f0cdb",
   "metadata": {},
   "source": [
    "#### Loading Human dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9aab89a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading the dataset, this dataset has two copies of the samples.The 2nd copy is equal to the 1st multiplied by a constant\n",
    "# unique to each sample.\n",
    "human_datamatrix_base = pd.read_excel('ST001082_AN001766_HD.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b7b73f2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2_Pooled_Sample_1</th>\n",
       "      <th>2_Pooled_Sample_1_2</th>\n",
       "      <th>P2_2</th>\n",
       "      <th>P2_3</th>\n",
       "      <th>P2_4</th>\n",
       "      <th>Psp_1</th>\n",
       "      <th>Psp_2</th>\n",
       "      <th>Psp_3</th>\n",
       "      <th>Psp_4</th>\n",
       "      <th>Psp_5</th>\n",
       "      <th>...</th>\n",
       "      <th>S67_3.1</th>\n",
       "      <th>S70_1.1</th>\n",
       "      <th>S70_2.1</th>\n",
       "      <th>S70_3.1</th>\n",
       "      <th>S72_1.1</th>\n",
       "      <th>S72_2.1</th>\n",
       "      <th>S72_3.1</th>\n",
       "      <th>S77_1.1</th>\n",
       "      <th>S77_2.1</th>\n",
       "      <th>S77_3.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metabolite_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Factors</th>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>Sample Type:Pooled</td>\n",
       "      <td>...</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "      <td>Sample Type:No Recurrence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95.06024579_7.857483333</th>\n",
       "      <td>2544.256904</td>\n",
       "      <td>2174.683995</td>\n",
       "      <td>2349.405324</td>\n",
       "      <td>2142.906787</td>\n",
       "      <td>2633.130278</td>\n",
       "      <td>2510.239989</td>\n",
       "      <td>3619.82127</td>\n",
       "      <td>2217.770816</td>\n",
       "      <td>2894.966994</td>\n",
       "      <td>20587.0328</td>\n",
       "      <td>...</td>\n",
       "      <td>4121.029059</td>\n",
       "      <td>3834.23691</td>\n",
       "      <td>4090.323428</td>\n",
       "      <td>3830.272712</td>\n",
       "      <td>3705.545567</td>\n",
       "      <td>3954.147516</td>\n",
       "      <td>3740.035077</td>\n",
       "      <td>3837.308267</td>\n",
       "      <td>4164.971893</td>\n",
       "      <td>4155.190276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71.06031304_0.064166667</th>\n",
       "      <td>153.7382834</td>\n",
       "      <td>2972.450721</td>\n",
       "      <td>951.0420238</td>\n",
       "      <td>2861.700714</td>\n",
       "      <td>3339.380877</td>\n",
       "      <td>299.8124672</td>\n",
       "      <td>783.1155979</td>\n",
       "      <td>3034.247021</td>\n",
       "      <td>3511.131943</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>8012.257751</td>\n",
       "      <td>853.7856117</td>\n",
       "      <td>6297.664545</td>\n",
       "      <td>7257.02521</td>\n",
       "      <td>845.5966076</td>\n",
       "      <td>6130.01575</td>\n",
       "      <td>7499.72554</td>\n",
       "      <td>796.3862056</td>\n",
       "      <td>5999.547509</td>\n",
       "      <td>7583.908921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71.06031992_9.984466667</th>\n",
       "      <td>292.2381888</td>\n",
       "      <td>1598.405492</td>\n",
       "      <td>638.9111168</td>\n",
       "      <td>1659.297423</td>\n",
       "      <td>1855.933656</td>\n",
       "      <td>382.0675008</td>\n",
       "      <td>948.3162562</td>\n",
       "      <td>1707.100565</td>\n",
       "      <td>1981.102653</td>\n",
       "      <td>62862.60997</td>\n",
       "      <td>...</td>\n",
       "      <td>4440.820011</td>\n",
       "      <td>662.2607137</td>\n",
       "      <td>3347.680062</td>\n",
       "      <td>4147.143726</td>\n",
       "      <td>694.8464783</td>\n",
       "      <td>3388.086581</td>\n",
       "      <td>4194.416655</td>\n",
       "      <td>650.7198089</td>\n",
       "      <td>3381.510189</td>\n",
       "      <td>4134.43322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126.1023924_4.038416667</th>\n",
       "      <td>347.8838312</td>\n",
       "      <td>125.0657553</td>\n",
       "      <td>260.8042864</td>\n",
       "      <td>119.6837125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>803.1992227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>126.2802051</td>\n",
       "      <td>0.001242692</td>\n",
       "      <td>582.9410185</td>\n",
       "      <td>...</td>\n",
       "      <td>322.3904954</td>\n",
       "      <td>4488.12769</td>\n",
       "      <td>3755.93028</td>\n",
       "      <td>1616.363518</td>\n",
       "      <td>1806.669855</td>\n",
       "      <td>1365.45848</td>\n",
       "      <td>495.4602756</td>\n",
       "      <td>1993.623747</td>\n",
       "      <td>1663.073748</td>\n",
       "      <td>1059.509693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588.4147684_2.0983</th>\n",
       "      <td>12264.30083</td>\n",
       "      <td>1935.998098</td>\n",
       "      <td>2697.901871</td>\n",
       "      <td>1761.340872</td>\n",
       "      <td>369.0957455</td>\n",
       "      <td>12243.27287</td>\n",
       "      <td>574.4747938</td>\n",
       "      <td>221.0599403</td>\n",
       "      <td>11.37397281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6225.892205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8491.983882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9425.946784</td>\n",
       "      <td>358.953733</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177.1270579_2.0983</th>\n",
       "      <td>4477.875444</td>\n",
       "      <td>317.9191822</td>\n",
       "      <td>2573.952315</td>\n",
       "      <td>69.97337036</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5703.699795</td>\n",
       "      <td>NaN</td>\n",
       "      <td>477.9239747</td>\n",
       "      <td>960.7144488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5104.451502</td>\n",
       "      <td>NaN</td>\n",
       "      <td>248.6690155</td>\n",
       "      <td>4012.400432</td>\n",
       "      <td>189.415007</td>\n",
       "      <td>1938.392231</td>\n",
       "      <td>5601.351328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.42248559</td>\n",
       "      <td>4329.675354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374.2064688_2.0983</th>\n",
       "      <td>7056.085863</td>\n",
       "      <td>112.7816239</td>\n",
       "      <td>4444.488636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6690.765635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>462.1335736</td>\n",
       "      <td>572.3140362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6158.786604</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3297.509597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2256.379266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7587.036926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211.0531259_2.0983</th>\n",
       "      <td>531.2204529</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>960.3628518</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3691.872884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>247.9229798</td>\n",
       "      <td>2324.687339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3852.866117</td>\n",
       "      <td>4529.484962</td>\n",
       "      <td>0.725505126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206.0808246_2.0983</th>\n",
       "      <td>7176.296988</td>\n",
       "      <td>103.1776278</td>\n",
       "      <td>5587.177152</td>\n",
       "      <td>11.30429951</td>\n",
       "      <td>6.817855612</td>\n",
       "      <td>6298.994544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>319.4387404</td>\n",
       "      <td>2073.079254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2605.099027</td>\n",
       "      <td>28.84466839</td>\n",
       "      <td>1246.041872</td>\n",
       "      <td>5895.292753</td>\n",
       "      <td>NaN</td>\n",
       "      <td>140.9996109</td>\n",
       "      <td>1217.883423</td>\n",
       "      <td>193.9776142</td>\n",
       "      <td>352.7932301</td>\n",
       "      <td>6021.618225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14627 rows × 530 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          2_Pooled_Sample_1 2_Pooled_Sample_1_2  \\\n",
       "metabolite_name                                                   \n",
       "Factors                  Sample Type:Pooled  Sample Type:Pooled   \n",
       "95.06024579_7.857483333         2544.256904         2174.683995   \n",
       "71.06031304_0.064166667         153.7382834         2972.450721   \n",
       "71.06031992_9.984466667         292.2381888         1598.405492   \n",
       "126.1023924_4.038416667         347.8838312         125.0657553   \n",
       "...                                     ...                 ...   \n",
       "588.4147684_2.0983              12264.30083         1935.998098   \n",
       "177.1270579_2.0983              4477.875444         317.9191822   \n",
       "374.2064688_2.0983              7056.085863         112.7816239   \n",
       "211.0531259_2.0983              531.2204529                 NaN   \n",
       "206.0808246_2.0983              7176.296988         103.1776278   \n",
       "\n",
       "                                       P2_2                P2_3  \\\n",
       "metabolite_name                                                   \n",
       "Factors                  Sample Type:Pooled  Sample Type:Pooled   \n",
       "95.06024579_7.857483333         2349.405324         2142.906787   \n",
       "71.06031304_0.064166667         951.0420238         2861.700714   \n",
       "71.06031992_9.984466667         638.9111168         1659.297423   \n",
       "126.1023924_4.038416667         260.8042864         119.6837125   \n",
       "...                                     ...                 ...   \n",
       "588.4147684_2.0983              2697.901871         1761.340872   \n",
       "177.1270579_2.0983              2573.952315         69.97337036   \n",
       "374.2064688_2.0983              4444.488636                 NaN   \n",
       "211.0531259_2.0983                      NaN                 NaN   \n",
       "206.0808246_2.0983              5587.177152         11.30429951   \n",
       "\n",
       "                                       P2_4               Psp_1  \\\n",
       "metabolite_name                                                   \n",
       "Factors                  Sample Type:Pooled  Sample Type:Pooled   \n",
       "95.06024579_7.857483333         2633.130278         2510.239989   \n",
       "71.06031304_0.064166667         3339.380877         299.8124672   \n",
       "71.06031992_9.984466667         1855.933656         382.0675008   \n",
       "126.1023924_4.038416667                 NaN         803.1992227   \n",
       "...                                     ...                 ...   \n",
       "588.4147684_2.0983              369.0957455         12243.27287   \n",
       "177.1270579_2.0983                      NaN         5703.699795   \n",
       "374.2064688_2.0983                      NaN         6690.765635   \n",
       "211.0531259_2.0983                      NaN         960.3628518   \n",
       "206.0808246_2.0983              6.817855612         6298.994544   \n",
       "\n",
       "                                      Psp_2               Psp_3  \\\n",
       "metabolite_name                                                   \n",
       "Factors                  Sample Type:Pooled  Sample Type:Pooled   \n",
       "95.06024579_7.857483333          3619.82127         2217.770816   \n",
       "71.06031304_0.064166667         783.1155979         3034.247021   \n",
       "71.06031992_9.984466667         948.3162562         1707.100565   \n",
       "126.1023924_4.038416667                 NaN         126.2802051   \n",
       "...                                     ...                 ...   \n",
       "588.4147684_2.0983              574.4747938         221.0599403   \n",
       "177.1270579_2.0983                      NaN         477.9239747   \n",
       "374.2064688_2.0983                      NaN         462.1335736   \n",
       "211.0531259_2.0983                      NaN                 NaN   \n",
       "206.0808246_2.0983                      NaN         319.4387404   \n",
       "\n",
       "                                      Psp_4               Psp_5  ...  \\\n",
       "metabolite_name                                                  ...   \n",
       "Factors                  Sample Type:Pooled  Sample Type:Pooled  ...   \n",
       "95.06024579_7.857483333         2894.966994          20587.0328  ...   \n",
       "71.06031304_0.064166667         3511.131943                 NaN  ...   \n",
       "71.06031992_9.984466667         1981.102653         62862.60997  ...   \n",
       "126.1023924_4.038416667         0.001242692         582.9410185  ...   \n",
       "...                                     ...                 ...  ...   \n",
       "588.4147684_2.0983              11.37397281                 NaN  ...   \n",
       "177.1270579_2.0983              960.7144488                 NaN  ...   \n",
       "374.2064688_2.0983              572.3140362                 NaN  ...   \n",
       "211.0531259_2.0983                      NaN                 NaN  ...   \n",
       "206.0808246_2.0983              2073.079254                 NaN  ...   \n",
       "\n",
       "                                           S67_3.1                    S70_1.1  \\\n",
       "metabolite_name                                                                 \n",
       "Factors                  Sample Type:No Recurrence  Sample Type:No Recurrence   \n",
       "95.06024579_7.857483333                4121.029059                 3834.23691   \n",
       "71.06031304_0.064166667                8012.257751                853.7856117   \n",
       "71.06031992_9.984466667                4440.820011                662.2607137   \n",
       "126.1023924_4.038416667                322.3904954                 4488.12769   \n",
       "...                                            ...                        ...   \n",
       "588.4147684_2.0983                             NaN                6225.892205   \n",
       "177.1270579_2.0983                     5104.451502                        NaN   \n",
       "374.2064688_2.0983                     6158.786604                        NaN   \n",
       "211.0531259_2.0983                     3691.872884                        NaN   \n",
       "206.0808246_2.0983                     2605.099027                28.84466839   \n",
       "\n",
       "                                           S70_2.1                    S70_3.1  \\\n",
       "metabolite_name                                                                 \n",
       "Factors                  Sample Type:No Recurrence  Sample Type:No Recurrence   \n",
       "95.06024579_7.857483333                4090.323428                3830.272712   \n",
       "71.06031304_0.064166667                6297.664545                 7257.02521   \n",
       "71.06031992_9.984466667                3347.680062                4147.143726   \n",
       "126.1023924_4.038416667                 3755.93028                1616.363518   \n",
       "...                                            ...                        ...   \n",
       "588.4147684_2.0983                             NaN                        NaN   \n",
       "177.1270579_2.0983                     248.6690155                4012.400432   \n",
       "374.2064688_2.0983                             NaN                3297.509597   \n",
       "211.0531259_2.0983                     247.9229798                2324.687339   \n",
       "206.0808246_2.0983                     1246.041872                5895.292753   \n",
       "\n",
       "                                           S72_1.1                    S72_2.1  \\\n",
       "metabolite_name                                                                 \n",
       "Factors                  Sample Type:No Recurrence  Sample Type:No Recurrence   \n",
       "95.06024579_7.857483333                3705.545567                3954.147516   \n",
       "71.06031304_0.064166667                845.5966076                 6130.01575   \n",
       "71.06031992_9.984466667                694.8464783                3388.086581   \n",
       "126.1023924_4.038416667                1806.669855                 1365.45848   \n",
       "...                                            ...                        ...   \n",
       "588.4147684_2.0983                     8491.983882                        NaN   \n",
       "177.1270579_2.0983                      189.415007                1938.392231   \n",
       "374.2064688_2.0983                             NaN                        NaN   \n",
       "211.0531259_2.0983                             NaN                3852.866117   \n",
       "206.0808246_2.0983                             NaN                140.9996109   \n",
       "\n",
       "                                           S72_3.1                    S77_1.1  \\\n",
       "metabolite_name                                                                 \n",
       "Factors                  Sample Type:No Recurrence  Sample Type:No Recurrence   \n",
       "95.06024579_7.857483333                3740.035077                3837.308267   \n",
       "71.06031304_0.064166667                 7499.72554                796.3862056   \n",
       "71.06031992_9.984466667                4194.416655                650.7198089   \n",
       "126.1023924_4.038416667                495.4602756                1993.623747   \n",
       "...                                            ...                        ...   \n",
       "588.4147684_2.0983                             NaN                9425.946784   \n",
       "177.1270579_2.0983                     5601.351328                        NaN   \n",
       "374.2064688_2.0983                     2256.379266                        NaN   \n",
       "211.0531259_2.0983                     4529.484962                0.725505126   \n",
       "206.0808246_2.0983                     1217.883423                193.9776142   \n",
       "\n",
       "                                           S77_2.1                    S77_3.1  \n",
       "metabolite_name                                                                \n",
       "Factors                  Sample Type:No Recurrence  Sample Type:No Recurrence  \n",
       "95.06024579_7.857483333                4164.971893                4155.190276  \n",
       "71.06031304_0.064166667                5999.547509                7583.908921  \n",
       "71.06031992_9.984466667                3381.510189                 4134.43322  \n",
       "126.1023924_4.038416667                1663.073748                1059.509693  \n",
       "...                                            ...                        ...  \n",
       "588.4147684_2.0983                      358.953733                        NaN  \n",
       "177.1270579_2.0983                     25.42248559                4329.675354  \n",
       "374.2064688_2.0983                             NaN                7587.036926  \n",
       "211.0531259_2.0983                             NaN                        NaN  \n",
       "206.0808246_2.0983                     352.7932301                6021.618225  \n",
       "\n",
       "[14627 rows x 530 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_datamatrix = human_datamatrix_base.iloc[:-1, :-4] # Just select the rows corresponding to the dataset\n",
    "mz_list = human_datamatrix_base.iloc[:-1, -4] # Select column with list of m/z values\n",
    "\n",
    "human_datamatrix = human_datamatrix.set_index(human_datamatrix.columns[0]) # Set index as the metabolites name\n",
    "human_datamatrix = human_datamatrix.replace({0:np.nan}) # Replacing 0 values as missing values\n",
    "\n",
    "human_datamatrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c49b8d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeating m/z values with different retention times\n",
    "\n",
    "#abc = []\n",
    "#for i in list(mz_list):\n",
    "#    if i in abc:\n",
    "#        print(i)\n",
    "#    else:\n",
    "#        abc.append(i)\n",
    "#len(set(list(mz_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "43a3e520",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select one of the two copies of samples in the dataset by removing the samples ending with '.1'\n",
    "human_datamatrix = human_datamatrix[[i for i in human_datamatrix.columns if not i.endswith('.1')]]\n",
    "\n",
    "human_datamatrix = human_datamatrix.T # Transpose the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a571e010",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Sample Type:Blank',\n",
       " 'Sample Type:No Recurrence',\n",
       " 'Sample Type:Pooled',\n",
       " 'Sample Type:Recurrence'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Current sample labels. Sample will consist of 'Sample Type:No Recurrence' and 'Sample Type:Recurrence'\n",
    "hd_labels = list(human_datamatrix['Factors'])\n",
    "set(hd_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7b6bda40",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "135"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many 'No Recurrence' samples in the dataset\n",
    "hd_labels.count('Sample Type:No Recurrence')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1babbb25",
   "metadata": {},
   "source": [
    "### Blank Treatment\n",
    "\n",
    "- If blank_treatment = True:\n",
    "\n",
    "An average of the blanks will be subtracted to the remaining dataset. Missing values in the blanks were replaced by 0 to calculate the average of the blanks. Negative values that arise in the dataset from subtracting the blanks will be coded as 0/missing values in the dataset. With the BinSim pre-treatment, they will be coded as 0.\n",
    "\n",
    "Possible problem for BinSim: features that are very similar between samples might be slightly negative in one and be coded as 0 and slightly positive in the other and be coded as 1, despite being basically identical.\n",
    "\n",
    "- If blank_treatment = False:\n",
    "\n",
    "Blank samples are removed and not accounted for. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "293f2252",
   "metadata": {},
   "outputs": [],
   "source": [
    "blank_treatment = True\n",
    "#blank_treatment = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2946fb88",
   "metadata": {},
   "outputs": [],
   "source": [
    "if blank_treatment:\n",
    "    blanks = human_datamatrix[human_datamatrix['Factors'] == 'Sample Type:Blank'].iloc[:,1:] # Select blank samples\n",
    "    blanks = blanks.replace({np.nan:0}) \n",
    "    blanks = blanks.astype(float) # Get blank samples with floats (needed since datamatrix has strings)\n",
    "    blanks_average = blanks.mean() # Average of the blanks\n",
    "    blanks_average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "58569e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selecting the samples belonging to either the 'No Recurrence' or 'Recurrence' \n",
    "selection = []\n",
    "for i in human_datamatrix.loc[:, 'Factors']:\n",
    "    if i in ['Sample Type:No Recurrence', 'Sample Type:Recurrence']:\n",
    "        selection.append(True)\n",
    "    else:\n",
    "        selection.append(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bef0ae60",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_datamatrix = human_datamatrix[selection]\n",
    "# Creating the list of 'targets' (labels of samples) of the dataset with the 'No Recurrence' and 'Recurrence' classes \n",
    "hd_labels = []\n",
    "for i in list(human_datamatrix.iloc[:,0]):\n",
    "    if i == 'Sample Type:No Recurrence':\n",
    "        hd_labels.append('No Recurrence')\n",
    "    else:\n",
    "        hd_labels.append('Recurrence')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ae4d2c6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>metabolite_name</th>\n",
       "      <th>95.06024579_7.857483333</th>\n",
       "      <th>71.06031304_0.064166667</th>\n",
       "      <th>71.06031992_9.984466667</th>\n",
       "      <th>126.1023924_4.038416667</th>\n",
       "      <th>121.0717995_1.358816667</th>\n",
       "      <th>105.0697004_2.574533333</th>\n",
       "      <th>83.06023778_3.971433333</th>\n",
       "      <th>85.02828928_2.7925</th>\n",
       "      <th>85.0282975_2.440233333</th>\n",
       "      <th>88.05051567_0.095966667</th>\n",
       "      <th>...</th>\n",
       "      <th>370.0920791_2.0983</th>\n",
       "      <th>468.1567372_2.0983</th>\n",
       "      <th>727.5149425_2.0983</th>\n",
       "      <th>497.1417674_2.0983</th>\n",
       "      <th>400.1388094_2.0983</th>\n",
       "      <th>588.4147684_2.0983</th>\n",
       "      <th>177.1270579_2.0983</th>\n",
       "      <th>374.2064688_2.0983</th>\n",
       "      <th>211.0531259_2.0983</th>\n",
       "      <th>206.0808246_2.0983</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S10_1</th>\n",
       "      <td>5839.689646</td>\n",
       "      <td>2421.207099</td>\n",
       "      <td>1702.403305</td>\n",
       "      <td>1732.645968</td>\n",
       "      <td>11202.3156</td>\n",
       "      <td>13.77354446</td>\n",
       "      <td>15969.09873</td>\n",
       "      <td>11122.92882</td>\n",
       "      <td>76033.09098</td>\n",
       "      <td>15559.66213</td>\n",
       "      <td>...</td>\n",
       "      <td>549.5061406</td>\n",
       "      <td>1788.00406</td>\n",
       "      <td>237.4614192</td>\n",
       "      <td>3065.766426</td>\n",
       "      <td>170.0414581</td>\n",
       "      <td>1347.054595</td>\n",
       "      <td>479.0767276</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.58211389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S10_2</th>\n",
       "      <td>4925.936713</td>\n",
       "      <td>6760.239327</td>\n",
       "      <td>3927.481184</td>\n",
       "      <td>876.2460627</td>\n",
       "      <td>16464.75568</td>\n",
       "      <td>144.7380541</td>\n",
       "      <td>11235.281</td>\n",
       "      <td>17168.89308</td>\n",
       "      <td>50955.48348</td>\n",
       "      <td>17634.72493</td>\n",
       "      <td>...</td>\n",
       "      <td>2354.284435</td>\n",
       "      <td>428.9492324</td>\n",
       "      <td>6.856593556</td>\n",
       "      <td>126.8752621</td>\n",
       "      <td>6686.085356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5663.904982</td>\n",
       "      <td>11572.89831</td>\n",
       "      <td>1720.669345</td>\n",
       "      <td>3969.969384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S10_3</th>\n",
       "      <td>5530.737291</td>\n",
       "      <td>11234.04531</td>\n",
       "      <td>6138.510579</td>\n",
       "      <td>302.5141329</td>\n",
       "      <td>24441.42344</td>\n",
       "      <td>196.744245</td>\n",
       "      <td>6281.973729</td>\n",
       "      <td>23329.82436</td>\n",
       "      <td>38668.17711</td>\n",
       "      <td>17924.91677</td>\n",
       "      <td>...</td>\n",
       "      <td>1187.542997</td>\n",
       "      <td>8.16828659</td>\n",
       "      <td>4.535420845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6985.148407</td>\n",
       "      <td>5.695198924</td>\n",
       "      <td>5001.930045</td>\n",
       "      <td>9622.04692</td>\n",
       "      <td>1835.400914</td>\n",
       "      <td>3534.540622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S15_1</th>\n",
       "      <td>4345.962212</td>\n",
       "      <td>1647.020375</td>\n",
       "      <td>1087.357218</td>\n",
       "      <td>2044.425809</td>\n",
       "      <td>4820.244022</td>\n",
       "      <td>8.539063852</td>\n",
       "      <td>10824.9497</td>\n",
       "      <td>2858.654761</td>\n",
       "      <td>48787.91979</td>\n",
       "      <td>12293.8573</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2378.896459</td>\n",
       "      <td>578.7754145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155.6123097</td>\n",
       "      <td>2164.462213</td>\n",
       "      <td>936.4502838</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66.89873449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S15_2</th>\n",
       "      <td>4135.848719</td>\n",
       "      <td>7164.394072</td>\n",
       "      <td>3594.221384</td>\n",
       "      <td>1396.827349</td>\n",
       "      <td>6507.869956</td>\n",
       "      <td>56.61978795</td>\n",
       "      <td>8963.260007</td>\n",
       "      <td>6670.576787</td>\n",
       "      <td>49360.26082</td>\n",
       "      <td>23406.46182</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>437.5625196</td>\n",
       "      <td>11.89033556</td>\n",
       "      <td>1.767176652</td>\n",
       "      <td>1485.254178</td>\n",
       "      <td>73.86922602</td>\n",
       "      <td>1952.194542</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>282.3818514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S72_2</th>\n",
       "      <td>4042.390453</td>\n",
       "      <td>6266.816562</td>\n",
       "      <td>3463.69699</td>\n",
       "      <td>1395.930804</td>\n",
       "      <td>10517.7526</td>\n",
       "      <td>97.700184</td>\n",
       "      <td>17988.63512</td>\n",
       "      <td>4597.778125</td>\n",
       "      <td>24216.15305</td>\n",
       "      <td>23606.86636</td>\n",
       "      <td>...</td>\n",
       "      <td>7049.366359</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.910517997</td>\n",
       "      <td>1273.048903</td>\n",
       "      <td>3601.773523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1981.650461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3938.848802</td>\n",
       "      <td>144.146236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S72_3</th>\n",
       "      <td>3860.581052</td>\n",
       "      <td>7741.451008</td>\n",
       "      <td>4329.607912</td>\n",
       "      <td>511.4295756</td>\n",
       "      <td>15224.61293</td>\n",
       "      <td>594.5983793</td>\n",
       "      <td>9768.49416</td>\n",
       "      <td>7663.919663</td>\n",
       "      <td>17897.95381</td>\n",
       "      <td>12847.39403</td>\n",
       "      <td>...</td>\n",
       "      <td>5875.548618</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.355728465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15046.7641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5781.889838</td>\n",
       "      <td>2329.105172</td>\n",
       "      <td>4675.475887</td>\n",
       "      <td>1257.137318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S77_1</th>\n",
       "      <td>4741.732906</td>\n",
       "      <td>984.0884324</td>\n",
       "      <td>804.0895638</td>\n",
       "      <td>2463.50584</td>\n",
       "      <td>5199.447492</td>\n",
       "      <td>8.686503713</td>\n",
       "      <td>10287.69548</td>\n",
       "      <td>2550.038217</td>\n",
       "      <td>36939.60819</td>\n",
       "      <td>17136.32435</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2098.000087</td>\n",
       "      <td>3969.01007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81.4497374</td>\n",
       "      <td>11647.5714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.896501216</td>\n",
       "      <td>239.696676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S77_2</th>\n",
       "      <td>4153.731532</td>\n",
       "      <td>5983.356024</td>\n",
       "      <td>3372.384222</td>\n",
       "      <td>1658.585471</td>\n",
       "      <td>7551.248853</td>\n",
       "      <td>26.64915706</td>\n",
       "      <td>8092.077171</td>\n",
       "      <td>4522.858963</td>\n",
       "      <td>32505.31818</td>\n",
       "      <td>23193.57557</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>349.8420486</td>\n",
       "      <td>782.9663571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>486.1349482</td>\n",
       "      <td>357.9849943</td>\n",
       "      <td>25.35387578</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>351.8411172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S77_3</th>\n",
       "      <td>4004.531135</td>\n",
       "      <td>7308.931091</td>\n",
       "      <td>3984.52669</td>\n",
       "      <td>1021.093926</td>\n",
       "      <td>11603.34961</td>\n",
       "      <td>218.1597526</td>\n",
       "      <td>7231.994792</td>\n",
       "      <td>6320.658856</td>\n",
       "      <td>23194.78133</td>\n",
       "      <td>12536.85981</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.64118507</td>\n",
       "      <td>18.13739081</td>\n",
       "      <td>46.3256028</td>\n",
       "      <td>1381.508922</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4172.689722</td>\n",
       "      <td>7311.945681</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5803.286026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249 rows × 14626 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "metabolite_name 95.06024579_7.857483333 71.06031304_0.064166667  \\\n",
       "S10_1                       5839.689646             2421.207099   \n",
       "S10_2                       4925.936713             6760.239327   \n",
       "S10_3                       5530.737291             11234.04531   \n",
       "S15_1                       4345.962212             1647.020375   \n",
       "S15_2                       4135.848719             7164.394072   \n",
       "...                                 ...                     ...   \n",
       "S72_2                       4042.390453             6266.816562   \n",
       "S72_3                       3860.581052             7741.451008   \n",
       "S77_1                       4741.732906             984.0884324   \n",
       "S77_2                       4153.731532             5983.356024   \n",
       "S77_3                       4004.531135             7308.931091   \n",
       "\n",
       "metabolite_name 71.06031992_9.984466667 126.1023924_4.038416667  \\\n",
       "S10_1                       1702.403305             1732.645968   \n",
       "S10_2                       3927.481184             876.2460627   \n",
       "S10_3                       6138.510579             302.5141329   \n",
       "S15_1                       1087.357218             2044.425809   \n",
       "S15_2                       3594.221384             1396.827349   \n",
       "...                                 ...                     ...   \n",
       "S72_2                        3463.69699             1395.930804   \n",
       "S72_3                       4329.607912             511.4295756   \n",
       "S77_1                       804.0895638              2463.50584   \n",
       "S77_2                       3372.384222             1658.585471   \n",
       "S77_3                        3984.52669             1021.093926   \n",
       "\n",
       "metabolite_name 121.0717995_1.358816667 105.0697004_2.574533333  \\\n",
       "S10_1                        11202.3156             13.77354446   \n",
       "S10_2                       16464.75568             144.7380541   \n",
       "S10_3                       24441.42344              196.744245   \n",
       "S15_1                       4820.244022             8.539063852   \n",
       "S15_2                       6507.869956             56.61978795   \n",
       "...                                 ...                     ...   \n",
       "S72_2                        10517.7526               97.700184   \n",
       "S72_3                       15224.61293             594.5983793   \n",
       "S77_1                       5199.447492             8.686503713   \n",
       "S77_2                       7551.248853             26.64915706   \n",
       "S77_3                       11603.34961             218.1597526   \n",
       "\n",
       "metabolite_name 83.06023778_3.971433333 85.02828928_2.7925  \\\n",
       "S10_1                       15969.09873        11122.92882   \n",
       "S10_2                         11235.281        17168.89308   \n",
       "S10_3                       6281.973729        23329.82436   \n",
       "S15_1                        10824.9497        2858.654761   \n",
       "S15_2                       8963.260007        6670.576787   \n",
       "...                                 ...                ...   \n",
       "S72_2                       17988.63512        4597.778125   \n",
       "S72_3                        9768.49416        7663.919663   \n",
       "S77_1                       10287.69548        2550.038217   \n",
       "S77_2                       8092.077171        4522.858963   \n",
       "S77_3                       7231.994792        6320.658856   \n",
       "\n",
       "metabolite_name 85.0282975_2.440233333 88.05051567_0.095966667  ...  \\\n",
       "S10_1                      76033.09098             15559.66213  ...   \n",
       "S10_2                      50955.48348             17634.72493  ...   \n",
       "S10_3                      38668.17711             17924.91677  ...   \n",
       "S15_1                      48787.91979              12293.8573  ...   \n",
       "S15_2                      49360.26082             23406.46182  ...   \n",
       "...                                ...                     ...  ...   \n",
       "S72_2                      24216.15305             23606.86636  ...   \n",
       "S72_3                      17897.95381             12847.39403  ...   \n",
       "S77_1                      36939.60819             17136.32435  ...   \n",
       "S77_2                      32505.31818             23193.57557  ...   \n",
       "S77_3                      23194.78133             12536.85981  ...   \n",
       "\n",
       "metabolite_name 370.0920791_2.0983 468.1567372_2.0983 727.5149425_2.0983  \\\n",
       "S10_1                  549.5061406         1788.00406        237.4614192   \n",
       "S10_2                  2354.284435        428.9492324        6.856593556   \n",
       "S10_3                  1187.542997         8.16828659        4.535420845   \n",
       "S15_1                          NaN        2378.896459        578.7754145   \n",
       "S15_2                          NaN        437.5625196        11.89033556   \n",
       "...                            ...                ...                ...   \n",
       "S72_2                  7049.366359                NaN        0.910517997   \n",
       "S72_3                  5875.548618                NaN        0.355728465   \n",
       "S77_1                          NaN        2098.000087         3969.01007   \n",
       "S77_2                          NaN        349.8420486        782.9663571   \n",
       "S77_3                          NaN        33.64118507        18.13739081   \n",
       "\n",
       "metabolite_name 497.1417674_2.0983 400.1388094_2.0983 588.4147684_2.0983  \\\n",
       "S10_1                  3065.766426        170.0414581        1347.054595   \n",
       "S10_2                  126.8752621        6686.085356                NaN   \n",
       "S10_3                          NaN        6985.148407        5.695198924   \n",
       "S15_1                          NaN        155.6123097        2164.462213   \n",
       "S15_2                  1.767176652        1485.254178        73.86922602   \n",
       "...                            ...                ...                ...   \n",
       "S72_2                  1273.048903        3601.773523                NaN   \n",
       "S72_3                          NaN         15046.7641                NaN   \n",
       "S77_1                          NaN         81.4497374         11647.5714   \n",
       "S77_2                          NaN        486.1349482        357.9849943   \n",
       "S77_3                   46.3256028        1381.508922                NaN   \n",
       "\n",
       "metabolite_name 177.1270579_2.0983 374.2064688_2.0983 211.0531259_2.0983  \\\n",
       "S10_1                  479.0767276                NaN                NaN   \n",
       "S10_2                  5663.904982        11572.89831        1720.669345   \n",
       "S10_3                  5001.930045         9622.04692        1835.400914   \n",
       "S15_1                  936.4502838                NaN                NaN   \n",
       "S15_2                  1952.194542                NaN                NaN   \n",
       "...                            ...                ...                ...   \n",
       "S72_2                  1981.650461                NaN        3938.848802   \n",
       "S72_3                  5781.889838        2329.105172        4675.475887   \n",
       "S77_1                          NaN                NaN        0.896501216   \n",
       "S77_2                  25.35387578                NaN                NaN   \n",
       "S77_3                  4172.689722        7311.945681                NaN   \n",
       "\n",
       "metabolite_name 206.0808246_2.0983  \n",
       "S10_1                  41.58211389  \n",
       "S10_2                  3969.969384  \n",
       "S10_3                  3534.540622  \n",
       "S15_1                  66.89873449  \n",
       "S15_2                  282.3818514  \n",
       "...                            ...  \n",
       "S72_2                   144.146236  \n",
       "S72_3                  1257.137318  \n",
       "S77_1                   239.696676  \n",
       "S77_2                  351.8411172  \n",
       "S77_3                  5803.286026  \n",
       "\n",
       "[249 rows x 14626 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_datamatrix = human_datamatrix.iloc[:,1:]\n",
    "human_datamatrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "65097771",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_datamatrix = human_datamatrix.astype(float) # Passing the values from strings to floats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a9a65206",
   "metadata": {},
   "outputs": [],
   "source": [
    "if blank_treatment:\n",
    "    human_datamatrix = human_datamatrix.replace({np.nan:0}) - blanks_average\n",
    "    human_datamatrix[human_datamatrix<0] = 0\n",
    "    human_datamatrix = human_datamatrix.replace({0:np.nan})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "71bb1de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#human_datamatrix.columns = list(mz_list)[1:]\n",
    "#human_datamatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61528e66",
   "metadata": {},
   "source": [
    "Set list of masses to pass to Cytoscape and to build the MDiNs.\n",
    "\n",
    "'Neutralization' of the _m/z_ values by subtracting the mass of a proton to the _m/z_ peaks.\n",
    "\n",
    "4 peaks (2 pairs) with different retention time have the exact same _m/z_ values. These 2 pairs of peaks wil be treated as only 2 different peaks (instead of four) for MDiN building purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "be837647",
   "metadata": {},
   "outputs": [],
   "source": [
    "mz_list = mz_list[1:] - chemdict['H'][0] + electron_mass\n",
    "counts = human_datamatrix.count(axis=0)\n",
    "final_mz_list = list(mz_list[list(counts >= 2)])\n",
    "final_mz_list = set(final_mz_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9b6c4d98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>metabolite_name</th>\n",
       "      <th>71.06031304_0.064166667</th>\n",
       "      <th>71.06031992_9.984466667</th>\n",
       "      <th>126.1023924_4.038416667</th>\n",
       "      <th>121.0717995_1.358816667</th>\n",
       "      <th>105.0697004_2.574533333</th>\n",
       "      <th>83.06023778_3.971433333</th>\n",
       "      <th>85.02828928_2.7925</th>\n",
       "      <th>85.0282975_2.440233333</th>\n",
       "      <th>103.0539927_2.632416667</th>\n",
       "      <th>117.0695881_0.77655</th>\n",
       "      <th>...</th>\n",
       "      <th>370.0920791_2.0983</th>\n",
       "      <th>468.1567372_2.0983</th>\n",
       "      <th>727.5149425_2.0983</th>\n",
       "      <th>497.1417674_2.0983</th>\n",
       "      <th>400.1388094_2.0983</th>\n",
       "      <th>588.4147684_2.0983</th>\n",
       "      <th>177.1270579_2.0983</th>\n",
       "      <th>374.2064688_2.0983</th>\n",
       "      <th>211.0531259_2.0983</th>\n",
       "      <th>206.0808246_2.0983</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S10_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>767.728678</td>\n",
       "      <td>7664.437238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11885.760381</td>\n",
       "      <td>11034.772529</td>\n",
       "      <td>71411.834905</td>\n",
       "      <td>NaN</td>\n",
       "      <td>272.371357</td>\n",
       "      <td>...</td>\n",
       "      <td>549.506141</td>\n",
       "      <td>1788.004060</td>\n",
       "      <td>237.461419</td>\n",
       "      <td>3065.766426</td>\n",
       "      <td>170.041458</td>\n",
       "      <td>1347.054595</td>\n",
       "      <td>474.490232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.582114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S10_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12926.877318</td>\n",
       "      <td>128.638686</td>\n",
       "      <td>7151.942651</td>\n",
       "      <td>17080.736789</td>\n",
       "      <td>46334.227405</td>\n",
       "      <td>881.713011</td>\n",
       "      <td>470.898334</td>\n",
       "      <td>...</td>\n",
       "      <td>2354.284435</td>\n",
       "      <td>428.949232</td>\n",
       "      <td>6.856594</td>\n",
       "      <td>126.875262</td>\n",
       "      <td>6686.085356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5659.318486</td>\n",
       "      <td>11569.699041</td>\n",
       "      <td>1720.669345</td>\n",
       "      <td>3969.969384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S10_3</th>\n",
       "      <td>1546.96215</td>\n",
       "      <td>0.29663</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20903.545078</td>\n",
       "      <td>180.644877</td>\n",
       "      <td>2198.635380</td>\n",
       "      <td>23241.668069</td>\n",
       "      <td>34046.921035</td>\n",
       "      <td>1439.412175</td>\n",
       "      <td>101.298661</td>\n",
       "      <td>...</td>\n",
       "      <td>1187.542997</td>\n",
       "      <td>8.168287</td>\n",
       "      <td>4.535421</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6985.148407</td>\n",
       "      <td>5.695199</td>\n",
       "      <td>4997.343549</td>\n",
       "      <td>9618.847651</td>\n",
       "      <td>1835.400914</td>\n",
       "      <td>3534.540622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S15_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1079.508519</td>\n",
       "      <td>1282.365660</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6741.611351</td>\n",
       "      <td>2770.498470</td>\n",
       "      <td>44166.663715</td>\n",
       "      <td>155.791159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2378.896459</td>\n",
       "      <td>578.775415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155.612310</td>\n",
       "      <td>2164.462213</td>\n",
       "      <td>931.863788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66.898734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S15_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>431.910059</td>\n",
       "      <td>2969.991594</td>\n",
       "      <td>40.520419</td>\n",
       "      <td>4879.921658</td>\n",
       "      <td>6582.420496</td>\n",
       "      <td>44739.004745</td>\n",
       "      <td>168.413145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>437.562520</td>\n",
       "      <td>11.890336</td>\n",
       "      <td>1.767177</td>\n",
       "      <td>1485.254178</td>\n",
       "      <td>73.869226</td>\n",
       "      <td>1947.608046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>282.381851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S72_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>431.013514</td>\n",
       "      <td>6979.874238</td>\n",
       "      <td>81.600816</td>\n",
       "      <td>13905.296771</td>\n",
       "      <td>4509.621834</td>\n",
       "      <td>19594.896975</td>\n",
       "      <td>33.998957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7049.366359</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.910518</td>\n",
       "      <td>1273.048903</td>\n",
       "      <td>3601.773523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1977.063965</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3938.848802</td>\n",
       "      <td>144.146236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S72_3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11686.734568</td>\n",
       "      <td>578.499011</td>\n",
       "      <td>5685.155811</td>\n",
       "      <td>7575.763372</td>\n",
       "      <td>13276.697735</td>\n",
       "      <td>901.326825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5875.548618</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.355728</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15046.764100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5777.303342</td>\n",
       "      <td>2325.905903</td>\n",
       "      <td>4675.475887</td>\n",
       "      <td>1257.137318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S77_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1498.588550</td>\n",
       "      <td>1661.569130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6204.357131</td>\n",
       "      <td>2461.881926</td>\n",
       "      <td>32318.352115</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2098.000087</td>\n",
       "      <td>3969.010070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81.449737</td>\n",
       "      <td>11647.571400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.896501</td>\n",
       "      <td>239.696676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S77_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>693.668181</td>\n",
       "      <td>4013.370491</td>\n",
       "      <td>10.549789</td>\n",
       "      <td>4008.738822</td>\n",
       "      <td>4434.702672</td>\n",
       "      <td>27884.062105</td>\n",
       "      <td>164.837670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>349.842049</td>\n",
       "      <td>782.966357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>486.134948</td>\n",
       "      <td>357.984994</td>\n",
       "      <td>20.767380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>351.841117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S77_3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56.176636</td>\n",
       "      <td>8065.471248</td>\n",
       "      <td>202.060384</td>\n",
       "      <td>3148.656443</td>\n",
       "      <td>6232.502565</td>\n",
       "      <td>18573.525255</td>\n",
       "      <td>954.465181</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.641185</td>\n",
       "      <td>18.137391</td>\n",
       "      <td>46.325603</td>\n",
       "      <td>1381.508922</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4168.103226</td>\n",
       "      <td>7308.746412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5803.286026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249 rows × 12869 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "metabolite_name  71.06031304_0.064166667  71.06031992_9.984466667  \\\n",
       "S10_1                                NaN                      NaN   \n",
       "S10_2                                NaN                      NaN   \n",
       "S10_3                         1546.96215                  0.29663   \n",
       "S15_1                                NaN                      NaN   \n",
       "S15_2                                NaN                      NaN   \n",
       "...                                  ...                      ...   \n",
       "S72_2                                NaN                      NaN   \n",
       "S72_3                                NaN                      NaN   \n",
       "S77_1                                NaN                      NaN   \n",
       "S77_2                                NaN                      NaN   \n",
       "S77_3                                NaN                      NaN   \n",
       "\n",
       "metabolite_name  126.1023924_4.038416667  121.0717995_1.358816667  \\\n",
       "S10_1                         767.728678              7664.437238   \n",
       "S10_2                                NaN             12926.877318   \n",
       "S10_3                                NaN             20903.545078   \n",
       "S15_1                        1079.508519              1282.365660   \n",
       "S15_2                         431.910059              2969.991594   \n",
       "...                                  ...                      ...   \n",
       "S72_2                         431.013514              6979.874238   \n",
       "S72_3                                NaN             11686.734568   \n",
       "S77_1                        1498.588550              1661.569130   \n",
       "S77_2                         693.668181              4013.370491   \n",
       "S77_3                          56.176636              8065.471248   \n",
       "\n",
       "metabolite_name  105.0697004_2.574533333  83.06023778_3.971433333  \\\n",
       "S10_1                                NaN             11885.760381   \n",
       "S10_2                         128.638686              7151.942651   \n",
       "S10_3                         180.644877              2198.635380   \n",
       "S15_1                                NaN              6741.611351   \n",
       "S15_2                          40.520419              4879.921658   \n",
       "...                                  ...                      ...   \n",
       "S72_2                          81.600816             13905.296771   \n",
       "S72_3                         578.499011              5685.155811   \n",
       "S77_1                                NaN              6204.357131   \n",
       "S77_2                          10.549789              4008.738822   \n",
       "S77_3                         202.060384              3148.656443   \n",
       "\n",
       "metabolite_name  85.02828928_2.7925  85.0282975_2.440233333  \\\n",
       "S10_1                  11034.772529            71411.834905   \n",
       "S10_2                  17080.736789            46334.227405   \n",
       "S10_3                  23241.668069            34046.921035   \n",
       "S15_1                   2770.498470            44166.663715   \n",
       "S15_2                   6582.420496            44739.004745   \n",
       "...                             ...                     ...   \n",
       "S72_2                   4509.621834            19594.896975   \n",
       "S72_3                   7575.763372            13276.697735   \n",
       "S77_1                   2461.881926            32318.352115   \n",
       "S77_2                   4434.702672            27884.062105   \n",
       "S77_3                   6232.502565            18573.525255   \n",
       "\n",
       "metabolite_name  103.0539927_2.632416667  117.0695881_0.77655  ...  \\\n",
       "S10_1                                NaN           272.371357  ...   \n",
       "S10_2                         881.713011           470.898334  ...   \n",
       "S10_3                        1439.412175           101.298661  ...   \n",
       "S15_1                         155.791159                  NaN  ...   \n",
       "S15_2                         168.413145                  NaN  ...   \n",
       "...                                  ...                  ...  ...   \n",
       "S72_2                          33.998957                  NaN  ...   \n",
       "S72_3                         901.326825                  NaN  ...   \n",
       "S77_1                                NaN                  NaN  ...   \n",
       "S77_2                         164.837670                  NaN  ...   \n",
       "S77_3                         954.465181                  NaN  ...   \n",
       "\n",
       "metabolite_name  370.0920791_2.0983  468.1567372_2.0983  727.5149425_2.0983  \\\n",
       "S10_1                    549.506141         1788.004060          237.461419   \n",
       "S10_2                   2354.284435          428.949232            6.856594   \n",
       "S10_3                   1187.542997            8.168287            4.535421   \n",
       "S15_1                           NaN         2378.896459          578.775415   \n",
       "S15_2                           NaN          437.562520           11.890336   \n",
       "...                             ...                 ...                 ...   \n",
       "S72_2                   7049.366359                 NaN            0.910518   \n",
       "S72_3                   5875.548618                 NaN            0.355728   \n",
       "S77_1                           NaN         2098.000087         3969.010070   \n",
       "S77_2                           NaN          349.842049          782.966357   \n",
       "S77_3                           NaN           33.641185           18.137391   \n",
       "\n",
       "metabolite_name  497.1417674_2.0983  400.1388094_2.0983  588.4147684_2.0983  \\\n",
       "S10_1                   3065.766426          170.041458         1347.054595   \n",
       "S10_2                    126.875262         6686.085356                 NaN   \n",
       "S10_3                           NaN         6985.148407            5.695199   \n",
       "S15_1                           NaN          155.612310         2164.462213   \n",
       "S15_2                      1.767177         1485.254178           73.869226   \n",
       "...                             ...                 ...                 ...   \n",
       "S72_2                   1273.048903         3601.773523                 NaN   \n",
       "S72_3                           NaN        15046.764100                 NaN   \n",
       "S77_1                           NaN           81.449737        11647.571400   \n",
       "S77_2                           NaN          486.134948          357.984994   \n",
       "S77_3                     46.325603         1381.508922                 NaN   \n",
       "\n",
       "metabolite_name  177.1270579_2.0983  374.2064688_2.0983  211.0531259_2.0983  \\\n",
       "S10_1                    474.490232                 NaN                 NaN   \n",
       "S10_2                   5659.318486        11569.699041         1720.669345   \n",
       "S10_3                   4997.343549         9618.847651         1835.400914   \n",
       "S15_1                    931.863788                 NaN                 NaN   \n",
       "S15_2                   1947.608046                 NaN                 NaN   \n",
       "...                             ...                 ...                 ...   \n",
       "S72_2                   1977.063965                 NaN         3938.848802   \n",
       "S72_3                   5777.303342         2325.905903         4675.475887   \n",
       "S77_1                           NaN                 NaN            0.896501   \n",
       "S77_2                     20.767380                 NaN                 NaN   \n",
       "S77_3                   4168.103226         7308.746412                 NaN   \n",
       "\n",
       "metabolite_name  206.0808246_2.0983  \n",
       "S10_1                     41.582114  \n",
       "S10_2                   3969.969384  \n",
       "S10_3                   3534.540622  \n",
       "S15_1                     66.898734  \n",
       "S15_2                    282.381851  \n",
       "...                             ...  \n",
       "S72_2                    144.146236  \n",
       "S72_3                   1257.137318  \n",
       "S77_1                    239.696676  \n",
       "S77_2                    351.841117  \n",
       "S77_3                   5803.286026  \n",
       "\n",
       "[249 rows x 12869 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_datamatrix = transf.keep_atleast(human_datamatrix, minimum=2) # Keep features that appear in at least two samples\n",
    "human_datamatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca0b53d9",
   "metadata": {},
   "source": [
    "### Building datasets database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "862e29f0",
   "metadata": {},
   "source": [
    "Atomic mass of the most common naturally occurring isotope of the more important elements in metabolites and their isotopic abundances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2a314a17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target for grapevine 11-variety data sets\n",
      "['CAN', 'CAN', 'CAN', 'CS', 'CS', 'CS', 'LAB', 'LAB', 'LAB', 'PN', 'PN', 'PN', 'REG', 'REG', 'REG', 'RIP', 'RIP', 'RIP', 'RL', 'RL', 'RL', 'ROT', 'ROT', 'ROT', 'RU', 'RU', 'RU', 'SYL', 'SYL', 'SYL', 'TRI', 'TRI', 'TRI']\n",
      "------\n",
      "target for 4 yeast strains data set\n",
      "['WT', 'WT', 'WT', 'ΔGRE3', 'ΔGRE3', 'ΔGRE3', 'ΔENO1', 'ΔENO1', 'ΔENO1', 'ΔGLO1', 'ΔGLO1', 'ΔGLO1', 'ΔGLO2', 'ΔGLO2', 'ΔGLO2']\n",
      "------\n",
      "target for 2-class wild Vitis vs Vitis vinifera data set\n",
      "['wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera', 'wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera', 'vinifera', 'vinifera', 'vinifera', 'wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera']\n",
      "------\n",
      "target for human 2-class dataset\n",
      "['Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence', 'No Recurrence']\n"
     ]
    }
   ],
   "source": [
    "datasets = {}\n",
    "# From the alignments_new store\n",
    "\n",
    "# GD_neg_global2 (GDg2-)\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('all_1ppm_min2_neg').transpose()\n",
    "gd_labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "new_cols_df = data_df.columns + chemdict['H'][0] - electron_mass\n",
    "data_df.columns = new_cols_df\n",
    "\n",
    "datasets['GD_neg_global2'] = {'source': 'grapevine',\n",
    "                              'alignment': '1-2',\n",
    "                              'mode': '-',\n",
    "                              'name': 'GDg2-',\n",
    "                              'data': data_df,#.to_dict(orient='split'),\n",
    "                              'original': data_df,#.to_dict(orient='split'),\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "# GD_pos_global2 (GDg2+)\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('all_1ppm_min2_pos').transpose()\n",
    "labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "new_cols_df = data_df.columns - chemdict['H'][0] + electron_mass\n",
    "data_df.columns = new_cols_df\n",
    "\n",
    "datasets['GD_pos_global2'] = {'source': 'grapevine',\n",
    "                              'alignment': '1-2',\n",
    "                              'mode': '+',\n",
    "                              'name': 'GDg2+',\n",
    "                              'data': data_df,#.to_dict(orient='split'),\n",
    "                              'original': data_df,#.to_dict(orient='split'),\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "# GD_neg_class2 (GDc2-)\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('groups_1ppm_min2_all_1ppm_neg').transpose()\n",
    "labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "new_cols_df = data_df.columns + chemdict['H'][0] - electron_mass\n",
    "data_df.columns = new_cols_df\n",
    "\n",
    "datasets['GD_neg_class2'] = {'source': 'grapevine',\n",
    "                              'alignment': '2-1',\n",
    "                              'mode': '-',\n",
    "                              'name': 'GDc2-',\n",
    "                              'data': data_df,#.to_dict(orient='split'),\n",
    "                              'original': data_df,#.to_dict(orient='split'),\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "# GD_pos_class2 (GDc2+)\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('groups_1ppm_min2_all_1ppm_pos').transpose()\n",
    "labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "new_cols_df = data_df.columns - chemdict['H'][0] + electron_mass\n",
    "data_df.columns = new_cols_df\n",
    "\n",
    "datasets['GD_pos_class2'] = {'source': 'grapevine',\n",
    "                              'alignment': '2-1',\n",
    "                              'mode': '+',\n",
    "                              'name': 'GDc2+',\n",
    "                              'data': data_df,#.to_dict(orient='split'),\n",
    "                              'original': data_df,#.to_dict(orient='split'),\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "# YD (YD)\n",
    "yeast_labels = [item for item in yeast_classes for i in range(3)]\n",
    "\n",
    "datasets['YD'] = {'source': 'yeast',\n",
    "                            'alignment': '1-2',\n",
    "                            'mode': '+',\n",
    "                            'name': 'YD',\n",
    "                            'data': yeast_datamatrix,#.to_dict(orient='split'),\n",
    "                            'original': yeast_datamatrix,#.to_dict(orient='split'),\n",
    "                            'target': yeast_labels,\n",
    "                            'classes': list(pd.unique(yeast_labels))}\n",
    "\n",
    "# vitis_types (GD types)\n",
    "vitis_types = {'CAN': 'wild', 'RIP': 'wild', 'ROT': 'wild','RU': 'wild', 'LAB': 'wild',\n",
    "               'SYL': 'wild','REG': 'vinifera','CS': 'vinifera','PN': 'vinifera','RL': 'vinifera',\n",
    "               'TRI': 'vinifera'}\n",
    "\n",
    "gd_type_labels = [vitis_types[lbl] for lbl in gd_labels]\n",
    "\n",
    "datasets['vitis_types'] = {'source': 'grapevine',\n",
    "                            'alignment': '2-1',\n",
    "                            'mode': '-',\n",
    "                            'name': 'GD types',\n",
    "                            'data': datasets['GD_neg_class2']['original'],\n",
    "                            'original': datasets['GD_neg_class2']['original'],\n",
    "                            'target': gd_type_labels,\n",
    "                            'classes': list(pd.unique(gd_type_labels))}\n",
    "\n",
    "# HD (HD)\n",
    "datasets['HD'] = {'source': 'human',\n",
    "                  'alignment': '1-2',\n",
    "                  'mode': '+',\n",
    "                  'name': 'HD',\n",
    "                  'data': human_datamatrix,#.to_dict(orient='split'),\n",
    "                  'original': human_datamatrix,#.to_dict(orient='split'),\n",
    "                  'target': hd_labels,\n",
    "                  'classes': list(pd.unique(hd_labels))}\n",
    "\n",
    "\n",
    "print('target for grapevine 11-variety data sets')\n",
    "print(datasets['GD_neg_global2']['target'])\n",
    "print('------\\ntarget for 4 yeast strains data set')\n",
    "print(datasets['YD']['target'])\n",
    "print('------\\ntarget for 2-class wild Vitis vs Vitis vinifera data set')\n",
    "print(datasets['vitis_types']['target'])\n",
    "print('------\\ntarget for human 2-class dataset')\n",
    "print(datasets['HD']['target'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8c9dab99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fname = 'store_files/datasets.json'\n",
    "#with open(fname, \"w\", encoding='utf8') as write_file:\n",
    "    #json.dump(datasets, write_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "efcdb9c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# should not raise AssertionError:\n",
    "#assert_frame_equal(yeast_datamatrix, pd.DataFrame(datasets['YD']['data']['data'], index=datasets['YD']['data']['index'], columns=datasets['YD']['data']['columns']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "162669e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data_df = pd.HDFStore('alignments_new.h5').get('groups_1ppm_min2_all_1ppm_pos').transpose()\n",
    "#labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "#new_cols_df = data_df.columns - chemdict['H'][0] + electron_mass\n",
    "#data_df.columns = new_cols_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9882088c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# should not raise AssertionError:\n",
    "#assert_frame_equal(df, data_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11c53c20",
   "metadata": {},
   "source": [
    "### Writing a .csv file for each of the datasets with the list of masses that will be used to build the MDiNs (in Cytoscape's MetaNetter)\n",
    "\n",
    "The file needs to be correctly read by Cytoscape's MetaNetter (and Cytoscape) where the network will be built.\n",
    "\n",
    "For this, files with two columns both with the same list of masses (one index, one in the column) will be made. One will correspond to the network nodes names as a 'string'. The other will be an attribute of the corresponding node as a 'float' and will be used in the building process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ac82682d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for dskey, ds in datasets.items():\n",
    "    #if dskey.startswith('YD'):\n",
    "    #    ds['data'] = pd.DataFrame(ds['data']['data'], index=ds['data']['index'], columns=ds['data']['columns'])\n",
    "    #    ds['original'] = ds['data']\n",
    "        \n",
    "    if dskey.startswith('HD'):\n",
    "        #ds['data'] = pd.DataFrame(ds['data']['data'], index=ds['data']['index'], columns=ds['data']['columns'])\n",
    "        #ds['original'] = ds['data']\n",
    "        \n",
    "        # Specific mass lists for HD \n",
    "        # (two pairs of masses with same m/z values and different retention times are only treated as 2 different m/z)\n",
    "        final_mz_list = set(final_mz_list)\n",
    "        pd.DataFrame(final_mz_list, index=final_mz_list).to_csv('mass_data/MassList' + dskey + '.csv')\n",
    "        continue\n",
    "    \n",
    "    #else:\n",
    "        #df_idx = pd.MultiIndex.from_tuples(ds['data']['index'])\n",
    "        #df_data = pd.DataFrame(ds['data']['data'], index=df_idx, columns=ds['data']['columns'])\n",
    "        #df_data.index.set_names('label', level=0, inplace=True)\n",
    "        \n",
    "        #ds['data'] = df_data\n",
    "        #ds['original'] = df_data\n",
    "    \n",
    "    # Creating mass lists in .csv files for MDiN building\n",
    "    pd.DataFrame(ds['data'].columns, index=ds['data'].columns).to_csv('mass_data/MassList' + dskey + '.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d68ee29",
   "metadata": {},
   "source": [
    "### Dataset Characteristics\n",
    "\n",
    "Building a table with general characteristics about the 7 datasets studied."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d2bb0c17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># samples</th>\n",
       "      <th># features</th>\n",
       "      <th>features / sample (range)</th>\n",
       "      <th># classes</th>\n",
       "      <th>samples / class</th>\n",
       "      <th>% missing values</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Data set</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GDg2-</th>\n",
       "      <td>33</td>\n",
       "      <td>3629</td>\n",
       "      <td>658 (367-1002)</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>81.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDg2+</th>\n",
       "      <td>33</td>\n",
       "      <td>7026</td>\n",
       "      <td>1164 (355-2141)</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>83.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDc2-</th>\n",
       "      <td>33</td>\n",
       "      <td>3026</td>\n",
       "      <td>547 (338-919)</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>81.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDc2+</th>\n",
       "      <td>33</td>\n",
       "      <td>4565</td>\n",
       "      <td>824 (215-1670)</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>81.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YD</th>\n",
       "      <td>15</td>\n",
       "      <td>1893</td>\n",
       "      <td>646 (559-705)</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>65.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GD types</th>\n",
       "      <td>33</td>\n",
       "      <td>3026</td>\n",
       "      <td>547 (338-919)</td>\n",
       "      <td>2</td>\n",
       "      <td>15 Vitis vinifera, 18 Wild</td>\n",
       "      <td>81.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HD</th>\n",
       "      <td>249</td>\n",
       "      <td>12869</td>\n",
       "      <td>7936 (7057-8475)</td>\n",
       "      <td>2</td>\n",
       "      <td>114 Recurrence, 135 no Recurrence</td>\n",
       "      <td>38.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          # samples  # features features / sample (range)  # classes  \\\n",
       "Data set                                                               \n",
       "GDg2-            33        3629            658 (367-1002)         11   \n",
       "GDg2+            33        7026           1164 (355-2141)         11   \n",
       "GDc2-            33        3026             547 (338-919)         11   \n",
       "GDc2+            33        4565            824 (215-1670)         11   \n",
       "YD               15        1893             646 (559-705)          5   \n",
       "GD types         33        3026             547 (338-919)          2   \n",
       "HD              249       12869          7936 (7057-8475)          2   \n",
       "\n",
       "                            samples / class  % missing values  \n",
       "Data set                                                       \n",
       "GDg2-                                     3             81.86  \n",
       "GDg2+                                     3             83.43  \n",
       "GDc2-                                     3             81.91  \n",
       "GDc2+                                     3             81.94  \n",
       "YD                                        3             65.88  \n",
       "GD types         15 Vitis vinifera, 18 Wild             81.91  \n",
       "HD        114 Recurrence, 135 no Recurrence             38.33  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def characterize_dataset(dskey, ds):\n",
    "    \"Computes and returns some general characteristics of a dataset in a dictionary.\"\n",
    "\n",
    "    dataset_chrs = {}\n",
    "    \n",
    "    name = ds['name'] # Name of the dataset\n",
    "    n_samples, n_feats = ds['data'].shape\n",
    "    n_classes = len(ds['classes'])\n",
    "       \n",
    "    Feat_Sample = ds['data'].count(axis=1) # Nº Features in each sample\n",
    "    Min_Feat_Sample = str(Feat_Sample.min()) # Minimum nº Features in a sample\n",
    "    Max_Feat_Sample = str(Feat_Sample.max()) # Maximum nº Features in a sample\n",
    "    Average_Feat_Sample = Feat_Sample.mean() # Average nº Features in a sample\n",
    "    \n",
    "    avg_feat_per_sample = int(round(Average_Feat_Sample,0)) # Round\n",
    "    \n",
    "    Samp_Class = len(ds['target'])/len(ds['classes']) # Nº Sample per Class\n",
    "    if dskey == 'vitis_types':\n",
    "        Samp_Class = '15 Vitis vinifera, 18 Wild'\n",
    "        #Samp_Class = '15 $\\it{Vitis}$ $\\it{Vinifera}$, 18 Wild'\n",
    "    elif dskey == 'HD':\n",
    "        Samp_Class = '114 Recurrence, 135 no Recurrence'\n",
    "    else:\n",
    "        Samp_Class = str(int(Samp_Class))\n",
    "    \n",
    "    n_na = ds['data'].isna().sum().sum() # Nº of missing values in the dataset\n",
    "    \n",
    "    p_na = round(100.0 * n_na / (n_samples * n_feats), 2) # % of missing values in the dataset\n",
    "    \n",
    "    avg_na_per_feature = (n_samples - ds['data'].count(axis=0)).mean()\n",
    "    avg_na_per_feature = int(round(avg_na_per_feature, 0))\n",
    "    \n",
    "    return {'Data set': name,\n",
    "            '# samples': n_samples,\n",
    "            '# features': n_feats,\n",
    "            'features / sample (range)': f'{avg_feat_per_sample} ({Min_Feat_Sample}-{Max_Feat_Sample})',\n",
    "            '# classes': n_classes,\n",
    "            'samples / class':Samp_Class,\n",
    "            '% missing values': p_na,} \n",
    "            #'missing values / feature': avg_na_per_feature}\n",
    "\n",
    "data_characteristics = [characterize_dataset(dskey, ds) for dskey, ds in datasets.items()]\n",
    "data_characteristics = pd.DataFrame(data_characteristics).set_index('Data set')\n",
    "data_characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bfdb78a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data_characteristics.to_excel('paperimages/dataset_characteristics.xlsx', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de1f5e1f",
   "metadata": {},
   "source": [
    "## Sample Mass-Difference Network Construction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bce2dbcf",
   "metadata": {},
   "source": [
    "#### MDBs (Mass-Difference-based Building blocks) \n",
    "\n",
    "Building the **list of MDBs** to use when building the MDiNs.\n",
    "\n",
    "MDB - Mass-Difference-based Building blocks.\n",
    "\n",
    "The choice of this MDBs is then crucial in the network building process. Ideally, they should represent different 'simple' biochemical reactions that cover the most common and ubiquitous enzymatic and non-enzymatic reactions while also being a relatively small amount of reactions – a total of 15 were picked - and maintaining the metabolite formula charge neutrality. \n",
    "\n",
    "For example, to maintain neutrality, a phosphorylation would mean the overall addition of a PO3H – addition of a -PO3$^{2-}$ group + 2 H$^+$ (maintaining neutrality) to replace an H atom in a metabolite. All the MDBs chosen represent changes in metabolites of no more than 5 atoms and less than 80 Da (small size). Each MDB should represent a set of chemically known reactions and a change in every main element in metabolites (C, H, O, N, S and P) is represented by at least one of the MDBs. To fulfil these conditions, representative MDBs were searched using BRENDA (https://www.brenda-enzymes.org/). The groups chosen were the following:\n",
    "\n",
    "- CH2 (methylation) \n",
    "- O (oxygenation) \n",
    "- H2 (Hydrogenation)\n",
    "- O(N-H-) (Aminase): NH3(O-) - H2\n",
    "- PO3H (phosphorylation)\n",
    "- NH3(O-) (transaminases)\n",
    "- SO3 (sulphation)\n",
    "- CO (like formylation) \n",
    "- CO2 (carboxylation, decarboxylation)\n",
    "- CHOH (Hydroxymethylation) \n",
    "- NCH (formidoyltransferase)\n",
    "- CONH (carbamoyltransferase)\n",
    "- C2H2O (acetylation)\n",
    "- S (rare but an extra S reaction)\n",
    "- H2O\n",
    "\n",
    "There could be many other MDB representing other reactions that can also be included such as CN2H2 (amidinotransferases), COCH2COO (malonyl transferases), etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d02e752",
   "metadata": {},
   "source": [
    "Functions to transform formulas in string format to dictionary format and from calculate formulas (in dictionary format) exact masses.\n",
    "\n",
    "Example:\n",
    "\n",
    "    'String' format: 'C6H1206'\n",
    "    \n",
    "    'Dictionary' format: {'C':6, 'H':12, 'O':6}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a2bcc410",
   "metadata": {},
   "outputs": [],
   "source": [
    "def formula_process(formula):\n",
    "    \"\"\"Transforms a formula in string format into a dictionary.\"\"\"\n",
    "\n",
    "    # Empty dictionary to store the results\n",
    "    results = dict.fromkeys(['C','H','O','N','S','P','Cl','F'], 0)\n",
    "    count = ''\n",
    "    letter = None\n",
    "    minus = False\n",
    "    \n",
    "    # Run through the string\n",
    "    for i in range(len(formula)):\n",
    "        if formula[i].isupper(): # If i is an uppercase letter then it is an element\n",
    "            if letter: # Just to account for the first letter in the formula where letter is None \n",
    "                # Reached another letter, store previous results and reset count\n",
    "                if minus:\n",
    "                    results[letter] = - int(count or 1)\n",
    "                    minus = False\n",
    "                else:\n",
    "                    results[letter] = int(count or 1)\n",
    "                count = ''\n",
    "                \n",
    "            if i+1 < len(formula): # In case it's a two letter element such as Cl\n",
    "                if formula[i+1].islower(): # The second letter is always lower case\n",
    "                    letter = formula[i] + formula[i+1] # Store new 2 letter element\n",
    "                    continue\n",
    "                    \n",
    "            letter = formula[i] # Store new 1 letter element\n",
    "            \n",
    "        elif formula[i].isdigit():\n",
    "            count = count + formula[i] # If number, add number to count\n",
    "        \n",
    "        elif formula[i] == '-':\n",
    "            minus = True\n",
    "    \n",
    "    # Store results of the last letter\n",
    "    if minus:\n",
    "        results[letter] = - int(count or 1)\n",
    "        minus = False\n",
    "    else:\n",
    "        results[letter] = int(count or 1)\n",
    "                    \n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e1077c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getmass(c,h,o,n,s,p,cl,f):\n",
    "    \"Get the exact monoisotopic mass for any formula with the C,H,O,N,S,P,Cl and F elements only.\"\n",
    "    massC = chemdict['C'][0] * c\n",
    "    massH = chemdict['H'][0] * h\n",
    "    massO = chemdict['O'][0] * o\n",
    "    massN = chemdict['N'][0] * n\n",
    "    massS = chemdict['S'][0] * s\n",
    "    massP = chemdict['P'][0] * p\n",
    "    massCl = chemdict['Cl'][0] * cl\n",
    "    massF = chemdict['F'][0] * f \n",
    "\n",
    "    massTotal = massC + massH + massO + massN + massS + massP + massCl + massF\n",
    "\n",
    "    return massTotal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9351a69c",
   "metadata": {},
   "source": [
    "#### Making the transformation file with MDBs to use in Cytoscape\n",
    "\n",
    "The transformation file cwill be in a format accepted by Cytoscape's MetaNetter (.txt separated by tabs) to be considered as a list of transformations (with specific masses) to make the MDiN. 4 columns: 'Label','Formula','Mass','Selected' with mass being the most important.\n",
    "\n",
    "Formula changes will be put under the 'Label' column and the corresponding set of reactions they represent will be put under 'Formula'. This switch is to make the MDB Impact analysis done later down the line more legible (since 'formula changes' are easier to observe, quickly understood and more concise in relation to long-winded names for set of reactions).\n",
    "\n",
    "The 'Selected' column will be just a set of True for the MDBs to come in pre-selected in Cytoscape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fce51c04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Formula</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Selected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O(N-H-)</th>\n",
       "      <td>Deamination</td>\n",
       "      <td>0.984016</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH3(O-)</th>\n",
       "      <td>Transamination</td>\n",
       "      <td>1.031634</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2</th>\n",
       "      <td>Hydrogenation</td>\n",
       "      <td>2.015650</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH2</th>\n",
       "      <td>Methylation</td>\n",
       "      <td>14.015650</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <td>Oxygenation</td>\n",
       "      <td>15.994915</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2O</th>\n",
       "      <td>Condensation/Dehydration/Cyclization</td>\n",
       "      <td>18.010565</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCH</th>\n",
       "      <td>Formidoylation</td>\n",
       "      <td>27.010899</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CO</th>\n",
       "      <td>Formylation</td>\n",
       "      <td>27.994915</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHOH</th>\n",
       "      <td>Hydroxymethylation</td>\n",
       "      <td>29.002740</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>Tranfer of and -SH group</td>\n",
       "      <td>31.972071</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2H2O</th>\n",
       "      <td>Acetylation</td>\n",
       "      <td>42.010565</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONH</th>\n",
       "      <td>Carbamoylation</td>\n",
       "      <td>43.005814</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CO2</th>\n",
       "      <td>Carboxylation/Decarboxylation</td>\n",
       "      <td>43.989829</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SO3</th>\n",
       "      <td>Sulphation</td>\n",
       "      <td>79.956815</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PO3H</th>\n",
       "      <td>Phosphorylation</td>\n",
       "      <td>79.966331</td>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Formula       Mass Selected\n",
       "Label                                                            \n",
       "O(N-H-)                           Deamination   0.984016     true\n",
       "NH3(O-)                        Transamination   1.031634     true\n",
       "H2                              Hydrogenation   2.015650     true\n",
       "CH2                               Methylation  14.015650     true\n",
       "O                                 Oxygenation  15.994915     true\n",
       "H2O      Condensation/Dehydration/Cyclization  18.010565     true\n",
       "NCH                            Formidoylation  27.010899     true\n",
       "CO                                Formylation  27.994915     true\n",
       "CHOH                       Hydroxymethylation  29.002740     true\n",
       "S                    Tranfer of and -SH group  31.972071     true\n",
       "C2H2O                             Acetylation  42.010565     true\n",
       "CONH                           Carbamoylation  43.005814     true\n",
       "CO2             Carboxylation/Decarboxylation  43.989829     true\n",
       "SO3                                Sulphation  79.956815     true\n",
       "PO3H                          Phosphorylation  79.966331     true"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Prepare DataFrame\n",
    "trans_groups = pd.DataFrame(columns=['Label','Formula','Mass','Selected'])\n",
    "trans_groups = trans_groups.set_index('Label')\n",
    "\n",
    "# Chemical Formula transformations (MDBs chosen)\n",
    "MDBs = ['H2','CH2','CO2','O','CHOH','NCH','O(N-H-)','S','CONH','PO3H','NH3(O-)','SO3','CO', 'C2H2O', 'H2O']\n",
    "# Examples of the set of Reactions each MDB represents\n",
    "labels = ['Hydrogenation','Methylation','Carboxylation/Decarboxylation','Oxygenation','Hydroxymethylation',\n",
    "          'Formidoylation','Deamination','Tranfer of and -SH group','Carbamoylation','Phosphorylation','Transamination',\n",
    "          'Sulphation','Formylation','Acetylation','Condensation/Dehydration/Cyclization']\n",
    "\n",
    "for i in range(len(MDBs)): # Passing through each transformation/MDB\n",
    "\n",
    "    # Calculate the exact mass of the MDBs\n",
    "    form = formula_process(MDBs[i])\n",
    "    MDB_mass = getmass(form['C'],form['H'],form['O'],form['N'],form['S'],form['P'],form['Cl'],form['F'])\n",
    "    \n",
    "    # MDBs with removed atoms have to be done manually (formula_process function doesn't account for this since normal\n",
    "    # formulas don't have a negative amount of atoms of any elements)\n",
    "    #if MDBs[i] == 'O(-NH)':\n",
    "    #    MDB_mass = getmass(a['C'],-1,a['O'],-1,a['S'],a['P'],a['Cl'],a['F'])\n",
    "    #elif MDBs[i] == 'NH3(-O)':\n",
    "    #    MDB_mass = getmass(a['C'],a['H'],-1,a['N'],a['S'],a['P'],a['Cl'],a['F'])\n",
    "        \n",
    "    # Making the row for each transformation\n",
    "    string = MDBs[i]\n",
    "    trans_groups.loc[string] = (labels[i], MDB_mass, 'true')\n",
    "\n",
    "trans_groups = trans_groups.sort_values(by='Mass')\n",
    "trans_groups # Transformations DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "dcb0e94d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Saving the df as a .txt file in the described format\n",
    "trans_groups.to_csv('transgroups.txt', header=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f46bbc2",
   "metadata": {},
   "source": [
    "### Build the Mass-Difference Networks in Cytoscape using MetaNetter 2.0\n",
    "\n",
    "To build the network in Cytoscape we'll need:\n",
    "\n",
    "1) A list of neutral masses (with masses as float in node attributes).\n",
    "\n",
    "2) A list of allowed transformations with specific mass differences - list of MDBs that we build above.\n",
    "\n",
    "The transformation files and the dataset files were used in Cytoscape to build the MDiNs.\n",
    "\n",
    "Parameters: 1 ppm error allowed for edge establishment.\n",
    "\n",
    "The Yeast datasets already have the m/z 'buckets' that are already representing the neutral masses of the metabolites. To consider the neutral masses of the grapevine datasets, a simple transformation by adding or removing a proton (Hydrogen atom mass - electron mass), depending if the ionization mode is negative or positive, respectively, on the m/z peaks was made.\n",
    "\n",
    "\n",
    "Only one full network (for each benchmark dataset) is built. Then, subgraphs of them will be used to select every sample MDiN. This is the same as building an MDiN for each sample.\n",
    "\n",
    "The networks were exported in graphml format that networkX module can read.\n",
    "\n",
    "- Nodes have a standard number ID instead of the mass (stored as the attribute 'mass'). Other attributes stored are irrelevant. \n",
    "- Edges among the different attributes have a very useful attribute called 'Transformation' which stores which MDB of the list was used to establish the edge - will be used for MDB Impact analysis.\n",
    "- Finally, the graph is directed. Since reactions are bidireccional, they will be transformed to undirected graphs.\n",
    "\n",
    "Changes that will be made to the network:\n",
    "\n",
    "- Nodes will be identified by their masses.\n",
    "- Intensities of the node in each sample will be given in other notebooks to store for each specific subgraph."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "696665d8",
   "metadata": {},
   "source": [
    "## Data transformations and pre-treatments\n",
    "\n",
    "Each dataset is transformed by 6 combinations of missing value imputation and data pre-treatment methods (generating 6 data sets with different treatments applied).\n",
    "\n",
    "### Traditional and more established  intensity-based Pre-Treatments\n",
    "\n",
    "Missing value imputation is mandatory for the data sets since many statistical methods performed downstream can't handle missing values. \n",
    "\n",
    "Two missing value imputation procedures were applied before any intensity-based pre-treatments: \n",
    "\n",
    "- Missing values of a sample were replaced with a fifth of the minimum intensity value present in that sample of the data matrix - Limit of Detection type of imputation. This was performed with the `fillna_frac_sample_min`function below. \n",
    "\n",
    "- Missing values were imputed by Random Forest missing value imputation. Parameters: 10 trees, 100 nearest features considered, 0 minimum value, 1.0e10 maximum value and 2 max iterations, others were left as default. This was performed by using the scikit-learn Python module with `IterativeImputer`and `ExtraTreesRegressor` as estimator. In cases where this method was used, an extra **`_RF`** was added to the 'name' of the pre-treatment.\n",
    "\n",
    "#### Combinations of intensity-based pre-treatments made:\n",
    "\n",
    "- Ionly Treatment - Only Missing Value Imputation.\n",
    "\n",
    "- P Treatment - Missing Value Imputation and Pareto Scaling.\n",
    "\n",
    "- NP Treatment - Missing Value Imputation, Normalization and Pareto Scaling.\n",
    "\n",
    "- **NGP Treatment** - Missing Value Imputation, Normalization, Generalized Logarithmic Transformation and Pareto Scaling.\n",
    "\n",
    "For normalizations, Normalization by reference feature (Leucine Enkephalin) is used in all cases, except for the HD dataset where Probabilistic Quotient Normalization (PQN) is used with the mean of all samples acting as the reference sample.\n",
    "\n",
    "Note: Leucine Enkephalin peak (reference feature) is removed upon normalization. Order of pre-treatments is the order in which they were mentioned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "df7036d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def impute_RF(df, nearest_features=100, n_trees=50):\n",
    "    \"Random Forest Imputation of Missing Values.\"\n",
    "    rf_estimator = ExtraTreesRegressor(n_estimators=n_trees)\n",
    "    imputer = IterativeImputer(random_state=0, estimator=rf_estimator,\n",
    "                           n_nearest_features=nearest_features,\n",
    "                           min_value=0.0, max_value=1.0e10,\n",
    "                           max_iter=2,\n",
    "                           verbose=0)\n",
    "    imputed_data = imputer.fit_transform(df)\n",
    "    res = pd.DataFrame(imputed_data, index=df.index, columns=df.columns)\n",
    "    ncols = len(res.columns)\n",
    "    res = res.dropna(axis='columns', how='any')\n",
    "    ncols2 = len(res.columns)\n",
    "    if ncols > ncols2:\n",
    "        print(f'{ncols-ncols2} features dropped')\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "630bb8a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For missing value imputation based on constants relative to the minimum of each feature instead of the full dataset\n",
    "def fillna_frac_feat_min(df, fraction=0.2):\n",
    "    \"\"\"Set NaN to a fraction of the minimum value in each column of the DataFrame.\"\"\"\n",
    "\n",
    "    minimum = df.min(axis=0) * fraction\n",
    "    return df.fillna(minimum)\n",
    "\n",
    "# For missing value imputation based on constants relative to the minimum of each sample instead of the full dataset\n",
    "def fillna_frac_sample_min(df, fraction=0.2):\n",
    "    \"\"\"Set NaN to a fraction of the minimum value in each column of the DataFrame.\"\"\"\n",
    "\n",
    "    minimum = df.min(axis=1) * fraction\n",
    "    return df.T.fillna(minimum).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9748e672",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Performs all pre-treatment combinations mentioned\n",
    "def compute_transf(dataset, norm_ref=None, lamb=None):\n",
    "    \"Computes 6 combinations of pre-treatments and missing value imputations and returns after treatment datasets in a dict.\"\n",
    "    \n",
    "    data = dataset['data']\n",
    "    \n",
    "    # Imputation of Missing Values\n",
    "    #imputed = fillna_frac_feat_min(data, fraction=0.2)\n",
    "    imputed = fillna_frac_sample_min(data, fraction=0.2)\n",
    "    \n",
    "    # Imputation by RF\n",
    "    imputedRF = impute_RF(data, nearest_features=100, n_trees=10)\n",
    "    \n",
    "    \n",
    "    # Normalization by a reference feature\n",
    "    if norm_ref is not None:\n",
    "        norm = transf.normalize_ref_feature(imputed, norm_ref, remove=True)\n",
    "    else:\n",
    "        norm = imputed\n",
    "    \n",
    "    # Normalization by a reference feature RF\n",
    "    if norm_ref is not None:\n",
    "        normRF = transf.normalize_ref_feature(imputedRF, norm_ref, remove=True)\n",
    "    else:\n",
    "        normRF = imputedRF\n",
    "    \n",
    "    # Normalization by PQN for HD - 1/2 and RF   \n",
    "    if dataset['name'].startswith('HD'):\n",
    "        norm = transf.normalize_PQN(imputed, ref_sample='mean')\n",
    "        normRF = transf.normalize_PQN(imputedRF, ref_sample='mean')\n",
    "\n",
    "    \n",
    "    # Pareto Scaling and Generalized Logarithmic Transformation\n",
    "    P = transf.pareto_scale(imputed)\n",
    "    NP = transf.pareto_scale(norm)\n",
    "    NGP = transf.pareto_scale(transf.glog(norm, lamb=lamb))\n",
    "\n",
    "    # Pareto Scaling and Generalized Logarithmic Transformation\n",
    "    P_RF = transf.pareto_scale(imputedRF)\n",
    "    NP_RF = transf.pareto_scale(normRF)\n",
    "    NGP_RF = transf.pareto_scale(transf.glog(normRF, lamb=lamb))\n",
    "    \n",
    "    # Store results\n",
    "    dataset['Ionly'] = imputed\n",
    "    dataset['P'] = P\n",
    "    dataset['NP'] = NP\n",
    "    dataset['NGP'] = NGP\n",
    "    \n",
    "    dataset['Ionly_RF'] = imputedRF\n",
    "    dataset['P_RF'] = P_RF\n",
    "    dataset['NP_RF'] = NP_RF\n",
    "    dataset['NGP_RF'] = NGP_RF    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d636d16e",
   "metadata": {},
   "source": [
    "Human leucine enkephalin (Sigma Aldrich) is the reference feature (internal standard, [M+H]+ = 556.276575 Da or [M-H]- = 554.262022 Da) used for these datasets.\n",
    "\n",
    "Search in the grapevine data sets for the reference feature and confirm the reference feature in the yeast data sets with `find_closest_features` from metabolinks package.\n",
    "\n",
    "There is **no** reference feature in the HD dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f7f59d17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Theoretical mass for negative mode Leucine Enkephalin - 554.262022\n",
    "# Theoretical mass for positive mode Leucine Enkephalin - 556.276575\n",
    "Leu_Enk_neg = 554.262022 + chemdict['H'][0] - electron_mass\n",
    "Leu_Enk_pos = 556.276575 - chemdict['H'][0] + electron_mass\n",
    "\n",
    "# Reference Feature in the yeast dataset\n",
    "leu_enk_name = 555.2692975341"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8c4a3edb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "looking for reference in GD_neg_global2 ...\n",
      "Found ref feature 555.269298452291\n",
      "In data: 555.269296452291 delta = -2.000e-06\n",
      "\n",
      "looking for reference in GD_pos_global2 ...\n",
      "Found ref feature 555.269298547709\n",
      "In data: 555.269293547709 delta = -5.000e-06\n",
      "\n",
      "looking for reference in GD_neg_class2 ...\n",
      "Found ref feature 555.269298452291\n",
      "In data: 555.269296452291 delta = -2.000e-06\n",
      "\n",
      "looking for reference in GD_pos_class2 ...\n",
      "Found ref feature 555.269298547709\n",
      "In data: 555.269293547709 delta = -5.000e-06\n",
      "\n",
      "looking for reference in vitis_types ...\n",
      "Found ref feature 555.269298452291\n",
      "In data: 555.269296452291 delta = -2.000e-06\n",
      "\n",
      "looking for reference in YD ...\n",
      "Found ref feature 555.2692975341 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "query_datasets = [name for name in datasets if datasets[name]['source']=='grapevine']\n",
    "\n",
    "for name in query_datasets:\n",
    "    ds = datasets[name]\n",
    "    print(f'looking for reference in {name} ...')\n",
    "    ref_variable = Leu_Enk_neg if ds['mode'] == '-' else Leu_Enk_pos\n",
    "    closest = transf.find_closest_features(ds['data'], features=[ref_variable])\n",
    "    if closest[ref_variable] is not None:\n",
    "        print('Found ref feature', ref_variable)\n",
    "        delta = closest[ref_variable] - ref_variable\n",
    "        print(f'In data: {closest[ref_variable]} delta = {delta:.3e}\\n')\n",
    "\n",
    "query_datasets = [name for name in datasets if name.startswith('YD')]\n",
    "\n",
    "ref_variable = leu_enk_name\n",
    "for name in query_datasets:\n",
    "    ds = datasets[name]\n",
    "    print(f'looking for reference in {name} ...') \n",
    "    closest = transf.find_closest_features(ds['data'], features=[ref_variable])\n",
    "    if closest[ref_variable] is not None:\n",
    "        print('Found ref feature', ref_variable, '\\n')\n",
    "    else:\n",
    "        print('Ref feature not found\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b590441d",
   "metadata": {},
   "source": [
    "##### Apply the different pre-treatments and get the results in their respective dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f0908696",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying pre-processing transformations to data in GDg2- ..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\francisco\\Anaconda3\\lib\\site-packages\\sklearn\\impute\\_iterative.py:685: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.\n",
      "  warnings.warn(\"[IterativeImputer] Early stopping criterion not\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done! took 76.879 s\n",
      "Applying pre-processing transformations to data in GDg2+ ..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\francisco\\Anaconda3\\lib\\site-packages\\sklearn\\impute\\_iterative.py:685: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.\n",
      "  warnings.warn(\"[IterativeImputer] Early stopping criterion not\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done! took 145.294 s\n",
      "Applying pre-processing transformations to data in GDc2- ..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\francisco\\Anaconda3\\lib\\site-packages\\sklearn\\impute\\_iterative.py:685: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.\n",
      "  warnings.warn(\"[IterativeImputer] Early stopping criterion not\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done! took 58.544 s\n",
      "Applying pre-processing transformations to data in GDc2+ ..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\francisco\\Anaconda3\\lib\\site-packages\\sklearn\\impute\\_iterative.py:685: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.\n",
      "  warnings.warn(\"[IterativeImputer] Early stopping criterion not\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done! took 86.180 s\n",
      "Applying pre-processing transformations to data in YD ..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\francisco\\Anaconda3\\lib\\site-packages\\sklearn\\impute\\_iterative.py:685: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.\n",
      "  warnings.warn(\"[IterativeImputer] Early stopping criterion not\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done! took 33.743 s\n",
      "Applying pre-processing transformations to data in GD types ..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\francisco\\Anaconda3\\lib\\site-packages\\sklearn\\impute\\_iterative.py:685: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.\n",
      "  warnings.warn(\"[IterativeImputer] Early stopping criterion not\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done! took 57.882 s\n",
      "Applying pre-processing transformations to data in HD ..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\francisco\\Anaconda3\\lib\\site-packages\\sklearn\\impute\\_iterative.py:685: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.\n",
      "  warnings.warn(\"[IterativeImputer] Early stopping criterion not\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done! took 869.336 s\n"
     ]
    }
   ],
   "source": [
    "for name, ds in datasets.items():\n",
    "    dataset_name = datasets[name][\"name\"]\n",
    "    print(f'Applying pre-processing transformations to data in {dataset_name}', end=' ...')\n",
    "    start = perf_counter()\n",
    "    \n",
    "    if name.startswith('YD'):\n",
    "        ref_variable = leu_enk_name\n",
    "    elif name.startswith('HD'):\n",
    "        ref_variable = None\n",
    "    else:\n",
    "        ref_variable = Leu_Enk_neg if ds['mode'] == '-' else Leu_Enk_pos\n",
    "    \n",
    "    compute_transf(ds, norm_ref=ref_variable)\n",
    "    \n",
    "    end = perf_counter()\n",
    "\n",
    "    print(f'done! took {(end - start):.3f} s')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d85fcbb2",
   "metadata": {},
   "source": [
    "### Generate json files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d08cd907",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensure dir exists\n",
    "path = Path.cwd() / \"store_files\"\n",
    "path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "storepath = Path.cwd() / \"store_files\" / 'processed_data.h5'\n",
    "\n",
    "store = pd.HDFStore(storepath, complevel=9, complib=\"blosc:blosclz\")\n",
    "#pd.set_option('io.hdf.default_format','table')\n",
    "\n",
    "# keep json serializable values and store dataFrames in HDF store\n",
    "\n",
    "serializable = {}\n",
    "\n",
    "for dskey, dataset in datasets.items():\n",
    "    serializable[dskey] = {}\n",
    "    for key, value in dataset.items():\n",
    "        #print(dskey, key)\n",
    "        if isinstance(value, pd.DataFrame):\n",
    "            storekey = dskey + '_' + key\n",
    "            #print('-----', storekey)\n",
    "            store[storekey] = value\n",
    "            serializable[dskey][key] = f\"INSTORE_{storekey}\"\n",
    "        else:\n",
    "            serializable[dskey][key] = value\n",
    "store.close()\n",
    "            \n",
    "\n",
    "path = path / 'processed_data.json'\n",
    "with open(path, \"w\", encoding='utf8') as write_file:\n",
    "    json.dump(serializable, write_file)\n",
    "\n",
    "#serializable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f750a040",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
